
```{r loading, echo=FALSE}
# library(synExtra)
library(tidyverse)
library(biomaRt)
library(DESeq2)
library(here)
library(furrr)
library(synapser)

paste_ <- function(...) {
  paste(..., sep = "_")
}

synapser::synLogin()


# 
# syn <- synExtra::synDownloader( "~/data/AMP-AD/figs" )
```

# Reading data


```{r read_data}
data_sets <- tibble(name = c("cdk", "ad", "dub", "cdk_new"))
# data_sets <- tibble(name = c("cdk", "ad", "dub"))

data_sets <- data_sets %>%
  mutate(
    meta = map(name, ~read_csv(here("wrangled", .x, "meta.csv"))),
    counts = map(name, ~read_csv(here("wrangled", .x, "counts.csv.gz")))
  )
```

## Sample table

```{r example_meta_table}
ggsave(file.path("sample_matching", "metadata_head.pdf"), gridExtra::tableGrob(head(dplyr::arrange(data_sets$meta[[1]], method, cell_line, time, drug, concentration))))

```


## Planned analyses

* Fraction of zeros, plot per gene across samples, bulk vs. dge
* Correlation gene expression, stratified by expression level?
* Normalize counts with DESeq2
* UpsetR concordance zeros across cell-lines/treatments
* Weighted jaccard similarity for comparing similarity of gene sets between
  drugs
* Dispersion estimate DGE vs. bulk
* Concordance differential expression DGE vs. bulk, L1000 genes and not


## Normalizing counts

Using DESeq2 to normalize contact counts. Normalizing each batch separately.

Checked whether it makes a difference to normalize by using `median` or
`genefilter::shorth` as location function. `shorth` supposed to work better for
smaller counts (DGE?). Doesn't make any difference though, sticking with default
`median`.

```{r deseq_normalization}
download.file(
  "ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/locus_groups/protein-coding_gene.txt",
  here("raw_downloads", "hgnc_protein_coding.tsv")
)

protein_coding <- read_tsv(here("raw_downloads", "hgnc_protein_coding.tsv"))

samples_passing <- data_sets %>%
  mutate(
    meta = map2(
      counts, meta,
      function(counts, meta) {
        counts %>%
          group_by(sample_id) %>%
          summarize(total = sum(count)) %>%
          ungroup() %>%
          left_join(meta, by = "sample_id") %>%
          filter(
            case_when(
              method == "bulk" & total > 2e6 ~ TRUE,
              method == "dge" & total > 5e4 ~ TRUE,
              TRUE ~ FALSE
            )
          )
      }
    )
  ) %>%
  # Make sure only samples that are included in the metadata table are in count table
  # And only protein coding genes
  mutate(
    counts = map2(counts, meta, ~filter(.x, sample_id %in% .y$sample_id, symbol %in% protein_coding$symbol))
  )

data_sets_deseq <- samples_passing %>%
  # Split by method
  mutate(
    meta = map(meta, group_nest, method, .key = "meta", keep = TRUE)
  ) %>%
  unnest(meta) %>%
  # Prepare input meta and count table for DEseq2
  mutate(
    meta_deseq = map(
      meta,
      ~.x %>%
        # Generate unique id for each combination of exp parameters
        mutate(condition = paste_(method, if (exists("cell_line")) cell_line else NULL, drug, concentration, time)) %>%
        column_to_rownames("sample_id")
    )
  ) %>%
  mutate(
    counts_deseq = pmap(
      .,
      function(meta_deseq, counts, ...) {
        counts %>%
          dplyr::select(symbol, sample_id, count) %>%
          filter(sample_id %in% rownames(meta_deseq)) %>%
          spread(sample_id, count) %>%
          mutate_at(vars(-symbol), replace_na, replace = 0) %>%
          column_to_rownames("symbol") %>%
          as.matrix()
      }
    )
  )

write_rds(
  data_sets_deseq,
  here("deseq", "deseq_input.rds")
)
```

