
```{r loading}
# library(synExtra)
library(tidyverse)
library(biomaRt)
library(DESeq2)

paste_ <- function(...) {
  paste(..., sep = "_")
}
```

# Loading expression data


## CDK dataset

CDK4/6 inhibitor dataset collected by Caitlin Mills

* DGE Counts matrix:  syn17022198
* Drug identities for DGE: syn17022199
* RNAseq counts matrix (RPKM): syn17022202
* Drug identities for RNAseq: syn17022201

Bulk counts in folder

* CDK46_response_201608 folder syn10164318 
  + counts syn10164324
  + meta syn10155361
* CDK46_response_201701 folder syn10164319
  + counts syn10155359
  + meta syn10155362

```{r read_cdk_data}
cdk_bulk_meta <- readr::read_csv(file.path("wrangled", "cdk_bulk_meta.csv"))
cdk_bulk_counts <- readr::read_csv(file.path("wrangled", "cdk_bulk_counts.csv.gz"))
cdk_dge_meta <- readr::read_csv(file.path("wrangled", "cdk_dge_meta.csv"))
cdk_dge_counts <- readr::read_csv(file.path("wrangled", "cdk_dge_counts.csv.gz"))

cdk_meta_combined <- readr::read_csv(file.path("wrangled", "cdk_meta_combined.csv"))
cdk_count_combined <- readr::read_csv(file.path("wrangled", "cdk_count_combined.csv.gz"))
cdk_meta_matched_samples <- readr::read_rds(file.path("wrangled", "cdk_meta_matched_samples.rds"))

ensembl_gene_id_mapping <- readr::read_csv(file.path("wrangled", "ensembl_gene_id_mapping.csv"))
```

```{r stats_for_presentation}
ggsave(file.path("sample_matching", "cdk_metadata_head.pdf"), gridExtra::tableGrob(head(dplyr::arrange(cdk_meta_combined, method, cell_line, time, drug, concentration))))

```


* Fraction of zeros, plot per gene across samples, bulk vs. dge
* Correlation gene expression, stratified by expression level?
* Normalize counts with DESeq2
* UpsetR concordance zeros across cell-lines/treatments
* Weighted jaccard similarity for comparing similarity of gene sets between
  drugs
* Dispersion estimate DGE vs. bulk
* Concordance differential expression DGE vs. bulk, L1000 genes and not


Using DESeq2 to normalize contact counts.

Checking whether it makes a difference to normalize by using `median` or
`genefilter::shorth` as location function. `shorth` supposed to work better for
smaller counts (DGE?). Doesn't make any difference though, sticking with default
`median`.

```{r deseq_normalization}
cdk_meta_deseq <- cdk_meta_combined %>%
  # Generate unique id for each combination of exp parameters
  dplyr::mutate(condition = paste_(method, cell_line, drug, concentration, time)) %>%
  as.data.frame() %>%
  `rownames<-`(.$sample_id)

cdk_count_deseq <- cdk_count_combined %>%
  dplyr::select(-method, -symbol) %>%
  tidyr::spread(sample_id, count) %>%
  as.data.frame() %>%
  `rownames<-`(.$ensembl_gene_id) %>%
  dplyr::select(-ensembl_gene_id)

locfuncs <- list(
  "shorth" = genefilter::shorth,
  "median" = median
) 
deseq_all <- locfuncs %>%
  purrr::map(
    ~DESeq2::DESeqDataSetFromMatrix(
      countData = cdk_count_deseq,
      colData = cdk_meta_deseq,
      design = ~ condition
    ) %>%
      DESeq2::estimateSizeFactors(locfunc = .x)
  ) %>%
  set_names(names(locfuncs))
  
normalized_counts_all <- deseq_all %>%
  purrr::map(
    ~DESeq2::counts(.x, normalized = TRUE) %>%
      as.data.frame() %>%
      dplyr::mutate(gene_id = row.names(.x))
  ) %>%
  dplyr::bind_rows(.id = "locfunc")

norm_shorth_vs_median <- normalized_counts_all %>%
  tidyr::gather(key = "sample", value = "normcount", -locfunc, -gene_id) %>%
  tidyr::spread(locfunc, normcount)

p <- norm_shorth_vs_median %>%
  dplyr::filter(sample %in% sample(unique(.$sample), 9)) %>%
  ggplot(
    aes(median, shorth)
  ) +
    geom_point() +
    scale_y_log10() +
    scale_x_log10() +
    facet_wrap(~sample)
dir.create("QC", showWarnings = FALSE)
ggsave(file.path("QC", "norm_shorth_vs_media.pdf"), p)

# Since there's no difference in results of locfuncs, just using median
normalized_counts_all <- normalized_counts_all %>%
  dplyr::filter(locfunc == "median") %>%
  dplyr::select(-locfunc) %>%
  tidyr::gather(key = "sample_id", value = "count", -gene_id)

# Creating some useful subsets of samples with specific properties.
# MCF7 treated with Abemaciclibat is the only condition where DGE and bulk match
# perfectly over all parameters (see sample_matching/matched_sample_table.pdf)

subset_sample_ids_2_rep <- cdk_meta_deseq %>%
  dplyr::arrange(condition) %>%
  dplyr::group_by(condition) %>%
  dplyr::filter(n() > 1) %>%
  .$sample_id

subsets_sample_ids <- list(
  "dge" = dplyr::filter(cdk_meta_combined, method == "dge")$sample_id,
  "bulk" = dplyr::filter(cdk_meta_combined, method == "bulk")$sample_id,
  "dge_2_rep" = dplyr::filter(cdk_meta_combined, method == "dge")$sample_id %>%
    intersect(subset_sample_ids_2_rep),
  "bulk_2_rep" = dplyr::filter(cdk_meta_combined, method == "bulk")$sample_id %>%
    intersect(subset_sample_ids_2_rep),
  "matched" = dplyr::filter(cdk_meta_matched_samples, bulk > 0, dge > 0)$sample_ids %>%
    purrr::reduce(union),
  "matched_2_rep" = dplyr::filter(cdk_meta_matched_samples, bulk > 1, dge > 1)$sample_ids %>%
    purrr::reduce(union),
  "mcf7_abemaciclib" = dplyr::filter(cdk_meta_matched_samples, cell_line == "MCF7", drug %in% c("Abemaciclib", "DMSO"), bulk > 1, dge > 1)$sample_ids %>%
    purrr::reduce(union)
)
readr::write_rds(subsets_sample_ids, file.path("sample_matching", "sample_subsets_list.rds"))

# Creating DESeq2 objects fo each sample subset with a simple ~condition
# design formula


deseq <- subsets_sample_ids %>%
  purrr::map(
    ~DESeq2::DESeqDataSetFromMatrix(
      cdk_count_deseq %>%
        dplyr::select_at(dplyr::vars(dplyr::one_of(sort(.x)))),
      cdk_meta_deseq %>%
        dplyr::filter(sample_id %in% .x) %>%
        dplyr::arrange(sample_id),
      design = ~condition
    ) %>%
      DESeq2::DESeq()
  )
dir.create("deseq", showWarnings = FALSE)
readr::write_rds(deseq, file.path("deseq", "deseq_datasets.rds"))
```


```{r calculate_counts}
count_funcs <- list(
  raw = . %>%
    DESeq2::counts(normalized = FALSE) %>%
    as_tibble(rownames = "gene_id"),
  normalized = . %>%
    DESeq2::counts(normalized = TRUE) %>%
    as_tibble(rownames = "gene_id"),
  variance_stabilized = . %>%
    DESeq2::vst() %>%
    SummarizedExperiment::assay() %>%
    as_tibble(rownames = "gene_id")
)

counts <- deseq %>%
  map(
    function (de) {
      count_funcs %>%
        map(~.x(de)) %>%
        map(gather, key = "sample_id", value = "count", -gene_id) %>%
        bind_rows(.id = "normalization")
    }
  ) %>%
  bind_rows(.id = "comparison")
write_csv(counts, file.path("deseq", "count_data.csv.gz"))

```

```{r}

de_data <- des %>%
  purrr::map(
    ~SummarizedExperiment::rowData(.x) %>%
      tibble::as_tibble() %>%
      dplyr::mutate(ensembl_gene_id = names(SummarizedExperiment::rowRanges(.x)))
  )
purrr::iwalk(de_data, ~readr::write_csv(.x, paste0("de_data_", .y, ".csv.gz")))

disp_plots <- des %>%
  purrr::map(
    DESeq2::plotDispEsts
  )


comp_de_1 <- DESeq2::DESeqDataSetFromMatrix(
  countData = cdk_count_matrix[, colnames(cdk_count_matrix) %in% sample_lists[["comp"]]],
  colData = cdk_meta_combined[row.names(cdk_meta_combined) %in% sample_lists[["comp"]], ] %>%
    dplyr::mutate(condition = paste_(cell_line, drug, concentration, time)),
  design = ~ condition + method
) %>%
  DESeq2::DESeq()

mcf7_comp_sample_ids <- reps %>%
  dplyr::filter(cell_line == "MCF7") %>%
  .$sample_ids %>%
  purrr::reduce(union)

comp_de_MCF7 <- DESeq2::DESeqDataSetFromMatrix(
  countData = cdk_count_matrix[, colnames(cdk_count_matrix) %in% mcf7_comp_sample_ids],
  colData = cdk_meta_combined[row.names(cdk_meta_combined) %in% mcf7_comp_sample_ids, ] %>%
    dplyr::mutate(concentration = log10(concentration + 1)),
  design = ~ method + concentration + method:concentration
) %>%
  DESeq2::DESeq()

res_de_MCF7 <- DESeq2::resultsNames(comp_de_MCF7)[-1] %>%
  purrr::map(
    function (coef) {
      DESeq2::results(comp_de_MCF7, name = coef) %>%
        as.data.frame() %>%
        dplyr::mutate(coef = coef) #%>%
        # DESeq2::lfcShrink(res = ., dds = comp_de_MCF7, coef = coef, type = "apeglm")
    }
  ) %>%
  dplyr::bind_rows()

ma_plots <- res_de_MCF7 %>%
  dplyr::arrange(desc(padj)) %>%
  ggplot(aes(baseMean, log2FoldChange, color = (padj < .05))) +
    geom_point() +
    scale_x_log10() +
    facet_wrap(~coef, scales = "free_y", ncol = 1)
ggsave("conc_method_int.pdf", ma_plots)


neg_log_trans <- function(base = exp(1)) {
  force(base)
  trans <- function(x) -log(x, base)
  inv <- function(x) base^(-x)
  scales::trans_new(paste0("neg-log-", format(base)), trans, inv,
    scales::log_breaks(base = base),
    domain = c(-Inf, -1e-100)
  )
}

volcano_plots <- res_de_MCF7 %>%
  dplyr::arrange(desc(padj)) %>%
  ggplot(aes(log2FoldChange, padj, color = (padj < .05))) +
    geom_point() +
    scale_y_continuous(trans = neg_log_trans(10)) +
    facet_wrap(~coef, scales = "free_y", nrow = 1)
ggsave("conc_method_int_volcano.pdf", volcano_plots)


conc_effect_bulk_vs_dge <- DESeq2::results(comp_de_MCF7, contrast=list("methodbulk.concentration","methoddge.concentration")) %>%
  as.data.frame()
conc_effect_bulk_vs_dge_ma <-  conc_effect_bulk_vs_dge %>%
  dplyr::arrange(desc(padj)) %>%
  ggplot(aes(baseMean, log2FoldChange, color = (padj < .05))) +
    geom_point() +
    scale_x_log10()

residuals <- counts(comp_de_MCF7, normalized=TRUE) - t(t(assays(comp_de_MCF7)[["mu"]])/sizeFactors(comp_de_MCF7))

resultsNames(comp_de_MCF7)

test_result <- DESeq2::results(comp_de_MCF7, contrast = c(0, 0, .15, .075)) %>%
  as.data.frame() %>%
  dplyr::arrange(desc(padj)) %>%
  ggplot(aes(baseMean, log2FoldChange, color = (padj < .05))) +
    geom_point() +
    scale_x_log10()

comp_de_MCF7_reduced <- DESeq2::DESeq(comp_de_MCF7, test = "LRT", reduced = ~ method + concentration)

res_de_MCF7_reduced <- DESeq2::resultsNames(comp_de_MCF7_reduced) %>%
  purrr::map(
    function (coef) {
      DESeq2::results(comp_de_MCF7, name = coef) %>%
        as.data.frame() %>%
        dplyr::mutate(coef = coef) #%>%
        # DESeq2::lfcShrink(res = ., dds = comp_de_MCF7, coef = coef, type = "apeglm")
    }
  ) %>%
  dplyr::bind_rows()


coef
```


```{r distribution_counts}

cdk_dge_counts_hist <- cdk_dge_counts %>%
  dplyr::mutate(l1000 = symbol %in% l1000$symbol) %>%
  # Setting zero counts to 0.1 for the purpose of this histogram, so that they're
  # visible on log x-scale
  dplyr::mutate(count = ifelse(count < .1, .1, count)) %>%
  dplyr::mutate(count_log10 = log10(count))

hist_breaks <- c(-1, -0.8, -0.6, -0.4, -0.2, 0, 0.2, 0.4, 0.6, 0.8, 1, 1.2, 
1.4, 1.6, 1.8, 2, 2.2, 2.4, 2.6, 2.8, 3, 3.2, 3.4, 3.6, 3.8, 
4)
cdk_dge_counts_hist_data <- cdk_dge_counts_hist %>%
  dplyr::group_by(l1000) %>%
  dplyr::summarize(hist = list(hist(count_log10, breaks = hist_breaks, plot = FALSE))) %>%
  dplyr::mutate(hist = purrr::map(hist, ~tibble::tibble(bin_start = .x$breaks[-1], bin_end = head(.x$breaks, -1), count = .x$counts))) %>%
  tidyr::unnest(hist) %>%
  tidyr::spread(l1000, value = count) %>%
  dplyr::rename("l1000" = `TRUE`, "not_l1000" = `FALSE`) %>%
  dplyr::mutate(proportion_l1000 = l1000/(l1000 + not_l1000), bin_mid = .5*(bin_start + bin_end))

p <- ggplot(cdk_dge_counts_hist_data,
            aes(bin_mid, proportion_l1000)) +
  geom_col()

p <- ggplot(cdk_dge_counts_hist_data,
            aes(bin_mid, count)) +
  geom_col(aes(fill = l1000), position = "stack")

p <- ggplot(cdk_dge_counts_hist,
       aes(count, fill = l1000)) +
  geom_histogram() +
  scale_x_log10()



p <- ggplot(cdk_dge_counts_hist %>% dplyr::mutate(count = ifelse(count < .1, .1, count)),
       aes(count, fill = l1000)) +
  geom_histogram() +
  scale_x_log10()

gene_freqs <- seq(from = .1, to = 1, length.out = 10)
ucounts <- sort(unique(cdk_dge_counts_hist$count))

gene_freq_data <- cdk_dge_counts_hist %>%
  dplyr::arrange(count) %>%
  dplyr::group_by(well, l1000) %>%
  dplyr::summarize(d = list(ecdf(count))) %>%
  dplyr::mutate(freqs = purrr::map(d, function (d) {
    purrr::map(gene_freqs, function (freq) {
       data.frame(
         count = ucounts,
         above = d(ucounts) > freq
       )
    }) %>%
      rlang::set_names(gene_freqs) %>%
      dplyr::bind_rows(.id = "thresh")
  })) %>%
  dplyr::select(-d) %>%
  tidyr::unnest(freqs) %>%
  # tidyr::spread(key = well, value = above) %>%
  # tidyr::fill(-l1000, -thresh, -count, .direction = "up") %>%
  # tidyr::gather(key = well, value = frac, -l1000, -thresh, -count) %>%
  dplyr::group_by(l1000, thresh, count) %>%
  dplyr::summarize(frac_exp = sum(above)/n())

p <- ggplot(gene_freq_data, aes(count, frac_exp)) +
  geom_step(aes(color = thresh)) +
  facet_wrap(~l1000) +
  scale_x_log10()

p <- ggplot(gene_freq_data, aes(count, frac_exp)) +
  geom_step(aes(color = l1000)) +
  facet_wrap(~thresh) +
  scale_x_log10()
  


```


Performing PCA on the MCF7 Abemaciclib samples from bulk and DGE

```{r pca_cdk}

pca_plot <- function (data, meta, aes = ggplot2::aes(PC1, PC2), extra_layers = NULL, ...) {
  p <- prcomp(data, ...)
  pstats <- t(summary(p)$importance) %>%
    tibble::as_tibble(rownames = "component") %>%
    dplyr::rename(sdev = `Standard deviation`, prop_var = `Proportion of Variance`, cum_prop = `Cumulative Proportion`) %>%
    dplyr::mutate(component = factor(component, levels = unique(component))) %>%
    # Remove all components after cumsum reaches .999
    dplyr::filter(cumsum(dplyr::lag(cum_prop > .95, default = FALSE)) <= 1)
  ploadings <- p$x %>%
    tibble::as_tibble() %>%
    dplyr::bind_cols(meta)
  p_plot <- ggplot(ploadings, aes) +
    geom_point()
  if (!is.null(extra_layers))
    p_plot <- p_plot + extra_layers
  # var_plot <- ggplot(pstats, ggplot2::aes(x = 1, y = prop_var, fill = component)) +
  #   geom_col(position = "stack") +
  #   # geom_text(ggplot2::aes(y = cum_prop, label = prop_var), halign = 0, valign = 1) +
  #   coord_flip() +
  #   guides(fill = FALSE)
    # theme(legend.position = "bottom")
  # browser()
  var_table <- gridExtra::tableGrob(
    pstats %>%
      dplyr::select(component, prop_var) %>%
      dplyr::mutate(prop_var = formatC(prop_var * 100, digits = 3, format = "fg")) %>%
      tidyr::spread(component, prop_var),
    rows = NULL,
    theme = gridExtra::ttheme_default(base_size = 6)
  )
  patchwork::wrap_plots(p_plot, var_table, heights = c(5, 1), ncol = 1)
}

mcf7_pca_plot <- pca_plot(
  vs_counts[["comp"]] %>%
    dplyr::select_at(vars(one_of(sort(mcf7_comp_sample_ids)))) %>%
    as.matrix() %>%
    t(),
  cdk_meta_combined[row.names(cdk_meta_combined) %in% mcf7_comp_sample_ids, ] %>%
    dplyr::arrange(row.names(.)),
  aes(PC1, PC2, shape = method, color = concentration),
  extra_layers = list(
    scale_color_continuous(trans = "log10"),
    labs(title = "MCF7 Abemaciclib")
  )
)
ggsave("mcf7_pca_plot.pdf", mcf7_pca_plot, height = 6, width = 5)
```


```{r variance}



```
