---
title: "DESeq2 drug treatments vs control"
author: "Clemens Hug"
date: "3/15/2019"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(here)
library(synExtra)
library(qs)

synapser::synLogin()
syn <- synDownloader(here("data"))

paste_ <- function(...) {
  paste(..., sep = "_")
}
```

Calculating differentially expressed genes between drug treatment conditions
and control for DGE.

```{r read_data}
deseq_input <- syn("syn25292412") %>%
  qread()

```




```{r deseq_drugs_vs_control}
wd <- here("deseq", "treatment_vs_control")
dir.create(wd, recursive = TRUE)

chunked_deseq_input <- deseq_input %>%
  lspcheminf::chunk_df(100, seed = 1) %>%
  enframe(name = "chunk", value = "data") %>%
  mutate(
    input_file = file.path(wd, paste0("deseq_input_", chunk, ".qs")),
    output_file = file.path(wd, paste0("deseq_result_", chunk, ".qs"))
  ) %>%
  # Have to get rid of some spurious samples where there is only control and some
  # other controls, eg. LPS
  rowwise() %>%
  mutate(
    data = data %>%
      filter(map_lgl(samples, ~!all(.x[["drug_conc"]] == 0))) %>%
      list()
  )

pwalk(
  chunked_deseq_input,
  function(data, input_file, ...) {
    qsave(
      data,
      input_file,
      preset = "fast"
    )
  }
)

process_deseq <- function(input, output) {
  library(tidyverse)
  unloadNamespace("synapser")
  unloadNamespace("PythonEmbedInR")
  
  run_deseq_comp <- function(meta_deseq, counts_deseq, design, ...) {
    attr(design, ".Environment") <- baseenv()
    de <- DESeq2::DESeqDataSetFromMatrix(
      counts_deseq %>%
        {.[, order(colnames(.))]},
      meta_deseq %>%
        {.[order(rownames(.)), ]},
      design = design
    ) %>%
      DESeq2::DESeq()
    name <- DESeq2::resultsNames(de)[[2]]
    res <- DESeq2::results(de, name = name, alpha = 0.1)
    shrunk <- DESeq2::lfcShrink(de, coef = name, res = res, type = "apeglm")
    out_res <- as.data.frame(shrunk) %>%
      rownames_to_column("ensembl_gene_id") %>%
      as_tibble() %>%
      left_join(
        as.data.frame(res) %>%
          rownames_to_column("ensembl_gene_id") %>%
          select(ensembl_gene_id, log2FoldChange_MLE = log2FoldChange),
        by = "ensembl_gene_id"
      )
    list(
      deseq_object = de,
      results = out_res
    )
  }
  deseq_input <- qs::qread(input)
  res <- pmap(deseq_input, run_deseq_comp)
  qs::qsave(
    res %>%
      map("results"),
    output,
    preset = "fast"
  )
}

library(batchtools)

reg <- makeRegistry(
  file.dir = file.path(wd, paste0("registry_", gsub(" ", "_", Sys.time()))),
  seed = 1
)
# reg <- loadRegistry()

batchMap(
  fun = process_deseq,
  input = chunked_deseq_input[["input_file"]],
  output = chunked_deseq_input[["output_file"]]
)

job_table <- findJobs() %>%
  # Chunk jobs into a single array job
  mutate(chunk = 1)


submitJobs(
  job_table,
  resources = list(
    memory = "10gb",
    ncpus = 1L,
    partition = "short",
    walltime = 30*60,
    chunks.as.arrayjobs = TRUE,
    # For some reason these nodes fail to execute R because of an "illegal instruction"
    exclude = "compute-f-17-[09-25]"
  )
)

deseq_results <- chunked_deseq_input %>%
  ungroup() %>%
  mutate(
    data = map2(
      data, output_file,
      # ~.x[["results"]] <- read_rds(.y)
      function(data, output_file) {
        x <- qread(output_file)
        if (nrow(data) != length(x))
          browser()
        data[["results"]] <- x
        data
      }
    )
  ) %>%
  pull(data) %>%
  bind_rows()

qsave(
  deseq_results,
  here("deseq", "deseq_treatment_vs_control.qs"),
  preset = "high"
)

qsave(
  deseq_results %>%
    select(-counts_deseq),
  here("deseq", "deseq_treatment_vs_control_no_counts.qs"),
  preset = "high"
)

deseq_results_long <- deseq_results %>%
  select(
    -samples, -design, -meta_deseq, -counts_deseq
  ) %>%
  unnest(results)

data.table::fwrite(
  deseq_results_long,
  here("deseq", "deseq_treatment_vs_control_results.csv.gz")
)

```

```{r deseq_bulk}
meta_bulk <- syn("syn21644142") %>%
  read_csv()

counts_bulk <- syn("syn21644266") %>%
  read_rds()

meta_bulk_deseq <- meta_bulk %>%
  mutate(
    condition = paste(cells, drug, time, sep = "_"),
    control_condition = paste(cells, "control", time, sep = "_")
  ) %>%
  {
    controls <- filter(., drug == "control") %>%
      group_nest(condition, .key = "controls", keep = TRUE)
    samples <- filter(., drug != "control")
    samples %>%
      group_nest(cells, drug, time, condition, control_condition, .key = "samples", keep = TRUE) %>%
      inner_join(
        controls, by = c("control_condition" = "condition")
      ) %>%
      filter(
        map_lgl(samples, ~length(unique(.x[["drug_conc"]])) > 1),
        map_lgl(controls, ~nrow(.x) > 1)
      ) %>%
      rowwise() %>%
      mutate(
        meta = bind_rows(
          controls,
          samples
        ) %>%
          mutate(drug_conc_log = log10(drug_conc + 1)) %>%
          arrange(sample_id) %>%
          list()
      )
  }

input_deseq_bulk <- meta_bulk_deseq %>%
  select(-samples, -controls) %>%
  rowwise() %>%
  mutate(
    meta_deseq = meta %>%
      column_to_rownames("sample_id") %>%
      list(),
    counts = counts_bulk$counts[[1]] %>%
      filter(sample_id %in% meta[["sample_id"]]) %>%
      arrange(sample_id) %>%
      select(ensembl_gene_id, count, sample_id) %>%
      spread(sample_id, count) %>%
      column_to_rownames("ensembl_gene_id") %>%
      as.matrix() %>%
      list()
  )

run_deseq_comp <- function(meta_deseq, counts_deseq, design, ...) {
  attr(design, ".Environment") <- baseenv()
  de <- DESeq2::DESeqDataSetFromMatrix(
    counts_deseq %>%
      {.[, order(colnames(.))]},
    meta_deseq %>%
      {.[order(rownames(.)), ]},
    design = design
  ) %>%
    DESeq2::DESeq()
  name <- DESeq2::resultsNames(de)[[2]]
  res <- DESeq2::results(de, name = name, alpha = 0.05)
  shrunk <- DESeq2::lfcShrink(de, coef = name, res = res, type = "apeglm")
  out_res <- as.data.frame(shrunk) %>%
    rownames_to_column("ensembl_gene_id") %>%
    as_tibble() %>%
    left_join(
      as.data.frame(res) %>%
        rownames_to_column("ensembl_gene_id") %>%
        select(ensembl_gene_id, log2FoldChange_MLE = log2FoldChange),
      by = "ensembl_gene_id"
    )
  list(
    deseq_object = de,
    results = out_res
  )
}

unloadNamespace("synapser")
unloadNamespace("PythonEmbedInR")
deseq_bulk <- input_deseq_bulk %>%
  rowwise() %>%
  mutate(
    deseq = run_deseq_comp(meta_deseq, counts, ~drug_conc_log) %>%
      list()
  )


deseq_res_bulk <- deseq_bulk %>%
  select(-meta, -meta_deseq, -counts) %>%
  rowwise() %>%
  mutate(
    deseq = deseq[["results"]] %>%
      list()
  ) %>%
  ungroup()

write_rds(
  deseq_res_bulk,
  here("deseq", "deseq_treatment_vs_control_bulk.rds")
)

```



```{r synapse}
activity <- synapser::Activity(
  name = "Run DESeq2",
  used = c(
    "syn21558154"
  ),
  executed = "https://github.com/clemenshug/dge_comparison/blob/master/02_deseq2_drug_treatment_vs_control.Rmd"
)

c(
  here("deseq", "deseq_treatment_vs_control.qs"),
  here("deseq", "deseq_treatment_vs_control_no_counts.qs"),
  here("deseq", "deseq_treatment_vs_control_results.csv.gz"),
  here("deseq", "deseq_treatment_vs_control_bulk.rds")
) %>%
  synStoreMany(parentId = "syn21558153", activity = activity, forceVersion = FALSE)

#rsync -rv --exclude *.fq.gz --exclude fastq --exclude *.fastq.gz --exclude *.bam --exclude rapmap --exclude transcriptome 2018_02_steven_rodriguez_ad_repurposing transfer:/n/files/ImStor/sorger/data/rnaseq/lincs/dge_reanalysis_clemens
```

