---
title: "Comparison Connectivity Map"
author: "Clemens Hug"
date: "2/6/2019"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(cmapR)
library(patchwork)
library(egg)

paste_ <- function(...) {
  paste(..., sep = "_")
}

theme_set(theme_bw())
```


# Goal

One of the goals of DGE experiments is to characterize the impact of drug treatment
the transcriptome with a cheap and scalable experiment. CMap is one of the existing
resources with a large datase of drug treaments. CMap uses an assay called L1000
to measure expression of ~1000 selected genes, representative of the larger
transcriptome.

The goal of this analysis is to

1. Check if we can reliably detect the L1000 genes with DGE
2. Compare drug profiles obtained by DGE with L1000 CMap signatures


```{r loading}
cdk_meta_combined <- readr::read_csv(file.path("wrangled", "cdk_meta_combined.csv"))

ensembl_gene_id_mapping <- readr::read_csv(file.path("wrangled", "ensembl_gene_id_mapping.csv"))

sample_subsets_list <- readr::read_rds(file.path("sample_matching", "sample_subsets_list.rds"))

deseq <- readr::read_rds(file.path("deseq", "deseq_datasets.rds"))
counts <- readr::read_csv(file.path("deseq", "count_data.csv.gz"))

ensembl_gene_id_mapping <- readr::read_csv(file.path("wrangled", "ensembl_gene_id_mapping.csv")) 

l1000 <- readr::read_csv(file.path("wrangled", "l1000_gene_list.csv"))
l1000_ensembl_gene_ids <- l1000 %>%
  dplyr::left_join(ensembl_gene_id_mapping, by = c("gene_symbol" = "symbol")) %>%
  tidyr::drop_na(ensembl_gene_id) %>%
  .$ensembl_gene_id
```

## L1000 genes in DGE

First, using naive threshold to define "detectable" genes. Requiring at least
5 counts seems adequate.


```{r distribution_l1000_counts}
# Using only raw counts for this analysis
l1000_counts <- counts %>%
  filter(comparison %in% c("dge", "bulk"), normalization == "raw") %>%
  mutate(l1000 = ifelse(gene_id %in% l1000_ensembl_gene_ids, "l1000", "not_l1000"))

l1000_line_data <- l1000_counts %>%
  split(.$comparison) %>%
  map(
    function (df) {
      df %>%
        mutate(gene_id = fct_reorder(gene_id, count, .fun = median)) %>%
        group_by(gene_id, l1000) %>%
        summarize(quant = list(quantile(count, probs = c(0, .25, .5, .75, 1)))) %>%
        mutate(quant = map(quant, enframe, name = "quantile", value = "count")) %>%
        unnest(quant)
    }
  )

l1000_line_plot <- l1000_line_data %>%
  imap(
    ~ggplot(., aes(as.integer(gene_id), count + .1)) +
      geom_line(aes(color = quantile, group = quantile)) +
      facet_wrap(~l1000) +
      scale_y_log10() +
      ggtitle(.y)
  ) %>%
  wrap_plots(ncol = 1)
ggsave(file.path("cmap_overlap", "l1000_count_distribution.pdf"), l1000_line_plot)
```



```{r detectable_l1000}

detectable_threshold <- 5
fraction_detectable <- l1000_counts %>%
  mutate(detectable = count >= detectable_threshold) %>%
  group_by(comparison, sample_id, l1000) %>%
  summarize(n_detectable = sum(detectable), fraction_detectable = n_detectable/n())

fraction_detectable_plot <- fraction_detectable %>%
  ggplot(aes(comparison, fraction_detectable, fill = l1000)) +
    geom_violin(draw_quantiles = c(.25, .5, .75)) +
    ggtitle("Fraction of genes â‰¥5")
ggsave(file.path("cmap_overlap", "fraction_detectable.pdf"), fraction_detectable_plot, width = 5, height = 5)
```

Also plotting how often each gene is detectable across samples.

```{r fraction_detectable_per_gene}
fraction_detectable_per_gene <- l1000_counts %>%
  mutate(detectable = count >= detectable_threshold) %>%
  group_by(comparison, gene_id, l1000) %>%
  summarize(fraction_detectable = sum(detectable)/n()) %>%
  ungroup()

# Output kernel density estimates for each input point defined by vectors x, y
# Adapted from grDevices::densCols to output densities instead of colors
density_2d <- function (x, y = NULL, nbin = 128, bandwidth) {
  xy <- xy.coords(x, y, setLab = FALSE)
  select <- is.finite(xy$x) & is.finite(xy$y)
  x <- cbind(xy$x, xy$y)[select, ]
  map <- grDevices:::.smoothScatterCalcDensity(x, nbin, bandwidth)
  mkBreaks <- function(u) u - diff(range(u))/(length(u) - 1)/2
  xbin <- cut(x[, 1], mkBreaks(map$x1), labels = FALSE)
  ybin <- cut(x[, 2], mkBreaks(map$x2), labels = FALSE)
  dens <- map$fhat[cbind(xbin, ybin)]
  dens[is.na(dens)] <- 0
  dens
}

density_2d_plot <- function (
  data, aesthetics,
  histogram_args = list(bins = 10), scatter_elements = list(), margin_elements = list()
) {
  marginals <- aesthetics[c("x", "y")] %>%
    imap(
      function (axis_var, axis) {
        p <- ggplot(data, aes(!!axis_var)) +
          rlang::exec(
            geom_histogram,
            aes(y = stat(count)/sum(stat(count))),
            !!!histogram_args
          ) +
          geom_step(aes(y = stat(y)), stat = "ecdf") +
          margin_elements
        if (axis == "y") p <- p + scale_y_reverse() + coord_flip()
        p
      }
    )
  density_data <- data %>%
    mutate(density = density_2d(!!aesthetics$x, !!aesthetics$y)) %>%
    arrange(density)
  p <- ggplot(density_data, aesthetics) +
    geom_point() +
    scatter_elements
  egg::ggarrange(
    plots = list(
      ggplot() + theme_void(),
      marginals$x,
      marginals$y,
      p
    ),
    widths = c(1, 4),
    heights = c(1, 3)
  )
}

fraction_detectable_per_gene_density_scatter <- fraction_detectable_per_gene %>%
  spread(comparison, fraction_detectable) %>%
  split(.$l1000) %>%
  imap(
    function(df, gene_set) {
      p <- density_2d_plot(
        df,
        aes(bulk, dge, color = density),
        histogram_args = list(
          breaks = seq(from = 0, to =1, length.out = 10),
          closed = "right"
        ),
        scatter_elements = list(
          coord_fixed(),
          lims(x = c(0, 1), y = c(0, 1)),
          labs(
            x = "RNA-seq (fraction of samples\nwith expression >5)",
            y = "DGE (fraction of samples\nwith expression >5)",
            color = "Density of\ngenes",
            title = gene_set
          ),
          scale_color_viridis_c()
        ),
        margin_elements = list(
          lims(x = c(0, 1)),
          labs(
            x = NULL,
            y = "Fraction\ngenes"
          )
        )
      )
      p
    }
  ) %>%
  wrap_plots(ncol = 1)
ggsave(file.path("cmap_overlap", "fraction_detectable_per_gene_density_scatter.pdf"), fraction_detectable_per_gene_density_scatter, width = 6, height = 8)

```



CMap data has been dowloaded from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742


```{r access_cmap_data}
dir.create("cmap_overlap")
cmap_path <- "~/data/DGE_comp/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx"
instance_meta <- readr::read_tsv("~/data/DGE_comp/GSE92742_Broad_LINCS_inst_info.txt")
gene_meta <- readr::read_tsv("~/data/DGE_comp/GSE92742_Broad_LINCS_gene_info.txt")

instance_meta_trans <- instance_meta %>%
  dplyr::mutate(drug = stringr::str_to_lower(pert_iname), cell_line = stringr::str_to_lower(cell_id))


shared_conditions <- cdk_meta_combined %>%
  dplyr::mutate_at(vars(drug, cell_line), stringr::str_to_lower) %>%
  dplyr::left_join(instance_meta_trans, by = c("cell_line", "drug", "time" = "pert_time"))

shared_drugs <- shared_conditions %>%
  dplyr::group_by(cell_line, drug) %>%
  dplyr::summarize(n_inst = sum(!is.na(inst_id))) %>%
  dplyr::arrange(desc(n_inst))

ggsave(
  file.path("cmap_overlap", "shared_conditions_table.pdf"),
  gridExtra::tableGrob(head(shared_drugs))
)


cmap_meta <- cmapR::read.gctx.meta(cmap_path, dimension = "row")

```

```{r fraction_tractable_deseq}
deseq_treatment_vs_control <- read_rds(file.path("deseq", "deseq_treatment_vs_control.rds"))

fraction_tractable <- deseq_treatment_vs_control %>%
  mutate(res_df = map(deseq_res, as.tibble, rownames = "gene_id")) %>%
  select(method, cell_line, drug, concentration, time, res_df) %>%
  unnest(res_df) %>%
  mutate(
    tractable = !is.na(padj),
    is_l1000 = gene_id %in% l1000_ensembl_gene_ids
  )

p <- ggplot(fraction_tractable, aes(baseMean)) +
  geom_histogram() +
  facet_wrap(method~tractable, scales = "free")
```
