---
title: "Comparison Connectivity Map"
author: "Clemens Hug"
date: "2/6/2019"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(cmapR)
library(patchwork)
library(egg)
library(broom)

paste_ <- function(...) {
  paste(..., sep = "_")
}

theme_set(theme_bw())
```


# Goal

One of the goals of DGE experiments is to characterize the impact of drug treatment
the transcriptome with a cheap and scalable experiment. CMap is one of the existing
resources with a large datase of drug treaments. CMap uses an assay called L1000
to measure expression of ~1000 selected genes, representative of the larger
transcriptome.

The goal of this analysis is to

1. Check if we can reliably detect the L1000 genes with DGE
2. Compare drug profiles obtained by DGE with L1000 CMap signatures


* Graph that shows L1000 genes on x-axis, sorted by value on y-axis, y-axis
  is fraction of librares gene was detected in
* Graph comparin L1000 represetnation between CDK and AD data sets. Both axes
  show fraction of libraries gene was detected in
* Functional enrichment comparison DGE vs L1000 for a well known drug treatment
  set, e.g. JAK inhibitors


```{r loading}

comp_res <- readr::read_rds(file.path("deseq", "deseq_treatment_vs_control_result.rds"))

l1000 <- readr::read_csv(file.path("wrangled", "l1000_gene_list.csv"))

drug_ids <- read_csv(file.path("wrangled", "lincs_drug_ids.csv"))
```

```{r read_data}
data_sets <- tibble(name = c("cdk", "ad"))

data_sets <- data_sets %>%
  mutate(
    meta = map(name, ~read_csv(file.path("wrangled", .x, "meta.csv"))),
    counts = map(name, ~read_csv(file.path("wrangled", .x, "counts.csv.gz")))
  )
```
```{r count_table}
counts <- data_sets %>%
  unnest(counts) %>%
  dplyr::select(name, symbol, sample_id, count) %>%
  left_join(data_sets$meta %>% bind_rows() %>% dplyr::select(sample_id, method), by = "sample_id")

```



## L1000 genes in DGE

First, using naive threshold to define "detectable" genes. Requiring at least
5 counts seems adequate.


```{r distribution_l1000_counts}
# Using only raw counts for this analysis
l1000_counts <- counts %>%
  mutate(l1000 = ifelse(symbol %in% l1000$gene_symbol, "l1000", "not_l1000"))

l1000_line_data <- l1000_counts %>%
  group_by(name, method) %>%
  group_modify(
    function (df, group) {
      df %>%
        mutate(symbol = fct_reorder(symbol, count, .fun = median)) %>%
        group_by(symbol, l1000) %>%
        summarize(quant = list(quantile(count, probs = c(0, .25, .5, .75, 1)))) %>%
        mutate(quant = map(quant, enframe, name = "quantile", value = "count")) %>%
        unnest(quant)
    }
  ) %>%
  ungroup()

l1000_line_plot <- l1000_line_data %>%
  group_nest(name, method) %>%
  mutate(
    plot = map(
      data,
      ~list(ggplot(.x, aes(as.integer(symbol), count + .1)) +
        geom_line(aes(color = quantile, group = quantile)) +
        facet_wrap(~l1000) +
        scale_y_log10() +
        ggtitle(paste(.y)))
    )
  )
ggsave(file.path("cmap_overlap", "l1000_count_distribution.pdf"), l1000_line_plot)
```



```{r detectable_l1000}

detectable_threshold <- 5
fraction_detectable <- l1000_counts %>%
  mutate(detectable = count >= detectable_threshold) %>%
  group_by(name, method, sample_id, l1000) %>%
  summarize(n_detectable = sum(detectable), fraction_detectable = n_detectable/n()) %>%
  ungroup()

fraction_detectable_plot <- fraction_detectable %>%
  ggplot(aes(name, fraction_detectable, fill = l1000)) +
    geom_violin(draw_quantiles = c(.25, .5, .75)) +
    facet_grid(vars(method)) +
    labs(
      x = "Dataset",
      y = "Fraction of samples where count â‰¥5",
      title = "Distribution of detectable genes"
    )
ggsave(file.path("cmap_overlap", "fraction_detectable.pdf"), fraction_detectable_plot, width = 5, height = 5)
```

Also plotting how often each gene is detectable across samples.

```{r fraction_detectable_per_gene}
fraction_detectable_per_gene <- l1000_counts %>%
  mutate(detectable = count >= detectable_threshold) %>%
  group_by(comparison, gene_id, l1000) %>%
  summarize(fraction_detectable = sum(detectable)/n()) %>%
  ungroup()

# Output kernel density estimates for each input point defined by vectors x, y
# Adapted from grDevices::densCols to output densities instead of colors
density_2d <- function (x, y = NULL, nbin = 128, bandwidth) {
  xy <- xy.coords(x, y, setLab = FALSE)
  select <- is.finite(xy$x) & is.finite(xy$y)
  x <- cbind(xy$x, xy$y)[select, ]
  map <- grDevices:::.smoothScatterCalcDensity(x, nbin, bandwidth)
  mkBreaks <- function(u) u - diff(range(u))/(length(u) - 1)/2
  xbin <- cut(x[, 1], mkBreaks(map$x1), labels = FALSE)
  ybin <- cut(x[, 2], mkBreaks(map$x2), labels = FALSE)
  dens <- map$fhat[cbind(xbin, ybin)]
  dens[is.na(dens)] <- 0
  dens
}

density_2d_plot <- function (
  data, aesthetics,
  histogram_args = list(bins = 10), scatter_elements = list(), margin_elements = list()
) {
  marginals <- aesthetics[c("x", "y")] %>%
    imap(
      function (axis_var, axis) {
        p <- ggplot(data, aes(!!axis_var)) +
          rlang::exec(
            geom_histogram,
            aes(y = stat(count)/sum(stat(count))),
            !!!histogram_args
          ) +
          geom_step(aes(y = stat(y)), stat = "ecdf") +
          margin_elements
        if (axis == "y") p <- p + scale_y_reverse() + coord_flip()
        p
      }
    )
  density_data <- data %>%
    mutate(density = density_2d(!!aesthetics$x, !!aesthetics$y)) %>%
    arrange(density)
  p <- ggplot(density_data, aesthetics) +
    geom_point() +
    scatter_elements
  egg::ggarrange(
    plots = list(
      ggplot() + theme_void(),
      marginals$x,
      marginals$y,
      p
    ),
    widths = c(1, 4),
    heights = c(1, 3)
  )
}

fraction_detectable_per_gene_density_scatter <- fraction_detectable_per_gene %>%
  spread(comparison, fraction_detectable) %>%
  split(.$l1000) %>%
  imap(
    function(df, gene_set) {
      p <- density_2d_plot(
        df,
        aes(bulk, dge, color = density),
        histogram_args = list(
          breaks = seq(from = 0, to =1, length.out = 10),
          closed = "right"
        ),
        scatter_elements = list(
          coord_fixed(),
          lims(x = c(0, 1), y = c(0, 1)),
          labs(
            x = "RNA-seq (fraction of samples\nwith expression >5)",
            y = "DGE (fraction of samples\nwith expression >5)",
            color = "Density of\ngenes",
            title = gene_set
          ),
          scale_color_viridis_c()
        ),
        margin_elements = list(
          lims(x = c(0, 1)),
          labs(
            x = NULL,
            y = "Fraction\ngenes"
          )
        )
      )
      p
    }
  ) %>%
  wrap_plots(ncol = 1)
ggsave(file.path("cmap_overlap", "fraction_detectable_per_gene_density_scatter.pdf"), fraction_detectable_per_gene_density_scatter, width = 6, height = 8)

```




## Genes tractable for differential expression by DESeq2

Another idea to quantify the proportion of L1000 genes that are detectable by
DGE is to perform pairwise comparisons of drug treatment and DMSO using DESeq2.
and quantifying how many genes are passing the DESeq2 filtering steps. Hope that
this is more practically relevant than arbitrary thresholds above.

DESeq2 does independent filtering based on the Bioconductor genefilter package 
to increase power of differential expression analysis. The default filter
takes into account the average expression of each gene across all replicates and
conditions. Genes that pass filter are "tractable" by DESeq2 for differential
expression.

```{r fraction_tractable_deseq}
deseq_treatment_vs_control <- read_rds(file.path("deseq", "deseq_treatment_vs_control.rds")) %>%
  mutate(comp_id = 1:n())

fraction_tractable <- deseq_treatment_vs_control %>%
  mutate(res_df = map(deseq_res, as.tibble, rownames = "gene_id")) %>%
  select(method, cell_line, drug, concentration, time, res_df) %>%
  unnest(res_df) %>%
  mutate(
    tractable = !is.na(padj),
    is_l1000 = ifelse(gene_id %in% l1000_ensembl_gene_ids, "l1000", "not_l1000")
  )

fraction_tractable_per_gene <- fraction_tractable %>%
  group_by(method, is_l1000, gene_id) %>%
  summarize(fraction_tractable = sum(tractable)/n()) %>%
  ungroup()

fraction_tractable_per_condition <- fraction_tractable %>%
  group_by(method, cell_line, drug, concentration, time) %>%
  summarize(fraction_tractable = sum(tractable)/n()) %>%
  ungroup()

fraction_tractable_per_gene_density_scatter <- fraction_tractable_per_gene %>%
  spread(method, fraction_tractable) %>%
  split(.$is_l1000) %>%
  imap(
    function(df, gene_set) {
      p <- density_2d_plot(
        df,
        aes(bulk, dge, color = density),
        histogram_args = list(
          breaks = seq(from = -.00001, to =1, length.out = 10),
          closed = "right"
        ),
        scatter_elements = list(
          coord_fixed(),
          lims(x = c(0, 1), y = c(0, 1)),
          labs(
            x = "RNA-seq (fraction of conditions)",
            y = "DGE (fraction of conditions)",
            color = "Density of\ngenes",
            title = gene_set
          ),
          scale_color_viridis_c()
        ),
        margin_elements = list(
          lims(x = c(0, 1)),
          labs(
            x = NULL,
            y = "Fraction\ngenes"
          )
        )
      )
      p
    }
  ) %>%
  wrap_plots(ncol = 1)
```

Unfortunately, there are simply too few bulk RNA-seq conditions, can't quantify
the proportion of conditions where gene is detectable in bulk RNA-seq.



CMap data has been dowloaded from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742


```{r access_cmap_data}
library(webchem)
library(cmapR)

dir.create("cmap_overlap", showWarnings = FALSE)
cmap_paths <- list(
  GSE92742 = "~/data/DGE_comp/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx",
  GSE70138 = "~/data/DGE_comp/GSE70138_Broad_LINCS_Level5_COMPZ_n118050x12328_2017-03-06.gctx"
)

signature_meta <- read_csv(file.path("wrangled", "cmap_signature_meta.csv.gz")) %>%
  mutate(pubchem_cid = as.character(pubchem_cid))

gene_meta <- read_csv(file.path("wrangled", "cmap_gene_meta.csv.gz"))

comp_res_pubchem_ids <- comp_res %>%
  distinct(drug) %>%
  mutate(
    pubchem_id = map(get_cid(str_to_lower(drug), from = "name"), as.character)
  )

comp_res_pubchem_ids_unnested <- comp_res_pubchem_ids %>%
  unnest(pubchem_id) %>%
  drop_na()

# # Fuzzy check if cell lines are called by the same name or slight variations
# cell_line_comp <- stringdist::stringdistmatrix(
#   cdk_meta_combined$cell_line %>% unique() %>% str_to_lower(),
#   signature_meta$cell_id %>% unique() %>% str_to_lower(),
#   useNames = "strings",
#   method = "osa"
# ) %>%
#   as_tibble(rownames = "cdk") %>%
#   gather("cmap", "distance", -cdk) %>%
#   filter(distance <= 2)

shared_conditions <- comp_res %>%
  select(name, method, cell_line, drug, concentration, time, condition) %>%
  group_by(name, method, drug, cell_line) %>%
  arrange(desc(concentration), desc(time), .by_group = TRUE) %>%
  slice(1) %>%
  ungroup() %>%
  mutate(drug = str_to_lower(drug)) %>%
  inner_join(comp_res_pubchem_ids_unnested, by = "drug") %>%
  inner_join(
    signature_meta %>% mutate(cell_id = str_to_lower(cell_id)),
    by = c("pubchem_id" = "pubchem_cid")
  ) %>%
  # Remove controls
  filter(!(pert_iname %in% c("DMSO", "dsmo"))) %>%
  # Remove matches with concentration too different (<50%)
  filter(abs(log10(concentration/pert_dose)) < 0.1760913) %>%
  # Remove matches with time too different (<50%)
  # All CDK data is 24h
  filter(abs(log10(time/pert_time)) < 0.1760913) %>%
  distinct()

# How many unique signatures?
unique_sigs <- shared_conditions %>%
  dplyr::count(sig_id) %>%
  dplyr::count(n)

# Now that we have the IDs of the signatures we can fetch them from the
# big GCTX matrix
cmap_data <- shared_conditions %>%
  group_nest(condition) %>%
  mutate(
    cmap_data = map(
      data,
      function(data) {
        parse.gctx(cmap_paths[[data$dataset[[1]]]], cid = unique(data$sig_id))@mat %>%
          as_tibble(rownames = "pr_gene_id") %>%
          gather("sig_id", "z_score", -pr_gene_id)
      }
    )
  )


cmap_data_prepped <- cmap_data %>%
  ungroup() %>%
  mutate(
    data = map2(data, cmap_data, ~left_join(.y, .x, by = "sig_id"))
  ) %>%
  select(-cmap_data) %>%
  unnest(data) %>%
  left_join(
    gene_meta %>%
      # Convert indicator variables to factor variable
      # showing if gene is directly measured by l1000, computationally inferred
      # with high fidelity (Best INFerred Genes "BING") or inferred with
      # lower confidence
      mutate(
        cmap_measurement = case_when(
          pr_is_lm == 1 ~ "landmark",
          pr_is_bing == 1 ~ "inferred_high_fidelity",
          TRUE ~ "inferred_low_fidelity"
        )
      ) %>%
      select(pr_gene_id, symbol, cmap_measurement) %>%
      mutate(pr_gene_id = as.character(pr_gene_id)),
    by = "pr_gene_id"
  )


```

```{r comparing_deseq_cmap_signatures}
cmap_data_vs_dge <- cmap_data_prepped %>%
  # select(comp_id, sig_id) %>%
  left_join(comp_res %>% unnest(result) %>% select(condition, symbol, log2FoldChange), by = c("condition", "symbol"))
  # select(comp_id, sig_id, cell_line, drug, concentration, time, deseq_res) %>%
  # mutate(deseq_res = map(deseq_res, as_tibble, rownames = "ensembl_gene_id")) %>%
  # unnest(deseq_res) %>%
  # left_join(ensembl_gene_id_mapping, by = "ensembl_gene_id") %>%
  # left_join(cmap_data, by = c("sig_id", "symbol")) %>%
  # drop_na(log2FoldChange, z_score)
```


```{r deseq_cmap_scatter}
deseq_cmap_scatter <- cmap_data_vs_dge %>%
  mutate(cmap_condition = paste(pert_iname, pert_dose, pert_time, cell_id, sep = "\n")) %>%
  group_nest(name, method, cell_line, drug, concentration, time, condition) %>%
  mutate(dge_condition = paste(drug, concentration, time, cell_line, sep = "\n")) %>%
  mutate(
    plot = pmap(
      .,
      function(dge_condition, data, ...) {
          ggplot(data, aes(z_score, log2FoldChange)) +
            # geom_hex() +
            stat_density_2d(
              aes(fill = stat(level)),
              geom = "polygon"
            ) +
            facet_grid(cmap_measurement ~ cmap_condition) +
            scale_fill_viridis_c(trans = "log10") +
            coord_fixed() +
            labs(
              x = "CMap z-score", y = "DGE log2FC",
              fill = "Density",
              title = dge_condition
            )
      }
    )
  )


# ggsave(
#   file.path("cmap_overlap", "cmap_zscore_vs_dge_l2fc_density.pdf"),
#   deseq_cmap_scatter,
#   width = 16, height = 5
# )
# 
# deseq_cmap_cor_test <- deseq_cmap_data %>%
#   mutate(condition = paste(cell_line, drug, concentration, time, sep = "_")) %>%
#   group_by(condition, cmap_measurement) %>%
#   do(
#     test = cor.test(.$z_score, .$log2FoldChange, method = "kendall", conf.int = TRUE) %>% broom::tidy()
#   ) %>%
#   unnest(test) %>%
#   mutate(p.value = sprintf("%.4f", p.value))

```


```{r deseq_cmap_fisher}
def <- tibble(p.value = as.double(NA))
deseq_cmap_fisher <- deseq_cmap_data %>%
  mutate(condition = paste(cell_line, drug, concentration, time, sep = "_")) %>%
  mutate(sig_deseq = padj < 0.05, sig_cmap = abs(z_score) > 1.96) %>%
  group_by(condition, cmap_measurement) %>%
  group_map(
    ~tibble(
      cont_table = list(table(.x[, c("sig_deseq", "sig_cmap")])),
      n_sig_deseq = sum(.x$sig_deseq, na.rm = TRUE),
      n_sig_cmap = sum(.x$sig_cmap, na.rm = TRUE)
    )
  ) %>%
  mutate(
    fisher_test = map(
      cont_table,
      ~possibly(fisher.test, NULL)(.x, conf.int = TRUE, conf.level = .95) %>%
        {if (!is.null(.)) tidy(.) else tibble(p.value = NA)}
    )
  ) %>%
  unnest(fisher_test) %>%
  ungroup()

deseq_cmap_fisher_table <- deseq_cmap_fisher %>%
  arrange(
    group_by(., condition) %>% mutate(p.value = min(p.value, na.rm = TRUE)) %>% pull(p.value),
    condition,
    cmap_measurement
  ) %>%
  select(condition, cmap_measurement, p.value, starts_with("n_sig")) %>%
  mutate(p.value = sprintf("%.03g", p.value)) %>%
  tableGrob(
    theme = ttheme_default(
      base_size = 10
      # core = list(fg_params = list(hjust = 0, x = .05))
    )
  )
ggsave(file.path("cmap_overlap", "cmap_overlap_fisher.pdf"), deseq_cmap_fisher_table, width = 7, height = 10)
```


Export DGE signatures as GMT files for analysis in CMAP

```{r morpheus_cmap}
gct_de_data <- deseq_treatment_vs_control %>%
  select(method, cell_line, drug, concentration, time, deseq, deseq_res) %>%
  mutate(
    sample_id = paste_(method, cell_line, drug, concentration, time),
    deseq_res = map2(
      deseq, deseq_res,
      ~.y %>%
        as.data.frame() %>%
        rownames_to_column("gene_id")
    )
  ) %>%
  unnest(deseq_res)

gene_ids_all_exp <- gct_de_data %>%
  group_by(sample_id) %>%
  summarize(gene_id_set = list(gene_id)) %>%
  pull(gene_id_set) %>%
  reduce(base::intersect) %>%
  sort()

gct_de_rdesc <- gene_meta %>%
  drop_na(symbol) %>%
  inner_join(ensembl_gene_id_mapping, by = "symbol") %>%
  filter(ensembl_gene_id %in% gene_ids_all_exp) %>%
  # Filter duplicate symbol/ensembl_id mappings, just 2
  group_by(ensembl_gene_id) %>%
  slice(1) %>%
  ungroup() %>%
  rename(id = ensembl_gene_id) %>%
  arrange(id) 

gct_de_cdesc <- deseq_treatment_vs_control %>%
  select(method, cell_line, drug, concentration, time) %>%
  mutate(id = paste_(method, cell_line, drug, concentration, time)) %>%
  arrange(id)

gct_de_data_mat <- gct_de_data %>%
  select(sample_id, gene_id, log2FoldChange) %>%
  filter(gene_id %in% gct_de_rdesc$id) %>%
  arrange(gene_id, sample_id) %>%
  spread(sample_id, log2FoldChange) %>%
  column_to_rownames("gene_id") %>%
  as.matrix()

gct_de_gct <- new(
  getClassDef("GCT", package = "cmapR"),
  mat = gct_de_data_mat, rid = gct_de_rdesc$id, cid = gct_de_cdesc$id,
  cdesc = as.data.frame(gct_de_cdesc), rdesc = as.data.frame(gct_de_rdesc)
)

cmapR::write.gct(
  gct_de_gct,
  file.path("cmap_overlap", "de_log2FC_unmoderated.gct")
)
```

```{r cmap_query}
de_cmap_sig <- comp_res %>%
  group_by(name, method, cell_line, drug, time) %>%
  arrange(desc(concentration), desc(time), .by_group = TRUE) %>%
  slice(1) %>%
  ungroup() %>%
  unnest(result) %>%
  genebabel::join_hgnc("symbol", "symbol", select_cols = "entrez_id") %>%
  drop_na(entrez_id) %>%
  filter(padj < 0.05) %>%
  mutate(
    direction = ifelse(log2FoldChange > 0, "up", "down")
  ) %>%
  filter(!grepl("bari", condition, ignore.case = TRUE)) %>%
  group_by(name, method, condition, direction) %>%
  filter(n() >= 15) %>%
  arrange(padj) %>%
  slice(1:min(150, n())) %>%
  summarize(
    gmt = list(list(head = condition[1], desc = condition[1], len = n(), entry = entrez_id))
  )


cmapR::write.gmt(de_cmap_sig$gmt, file.path("cmap_overlap", "de_gene_lists.gmt"))

de_cmap_sig %>%
  # Make sure we have up and down for each set, otherwise remove
  group_by(condition) %>%
  filter(n() == 2) %>%
  ungroup() %>%
  # Query tool only supports 25 querys at once, selecting first 25 with biggest gene sets
  spread(direction, gmt) %>%
  mutate(n_genes = map2_int(up, down, ~.x$len + .y$len)) %>%
  arrange(desc(n_genes)) %>%
  group_by(name) %>%
  slice(1:25) %>%
  ungroup() %>%
  gather("direction", "gmt", up, down) %>%
  group_by(name, direction) %>%
  group_nest() %>%
  pwalk(
    .,
    function(name, direction, data) {
      cmapR::write.gmt(data$gmt, file.path("cmap_overlap", paste0("de_gene_lists_", name, "_", direction, ".gmt")))
    }
  )

```


```{r cmap_query_result}


```


