---
title: "Comparison Connectivity Map"
author: "Clemens Hug"
date: "2/6/2019"
output: html_document
---

```{r setup, include=FALSE}
library(synExtra)
library(tidyverse)
library(cmapR)
library(here)
library(egg)
library(broom)

synapser::synLogin()
syn <- synExtra::synDownloader("data")

wd <- here("cmap_overlap")
dir.create(wd, showWarnings = FALSE)

theme_set(theme_bw())
```


# Goal

One of the goals of DGE experiments is to characterize the impact of drug treatment
the transcriptome with a cheap and scalable experiment. CMap is one of the existing
resources with a large datase of drug treaments. CMap uses an assay called L1000
to measure expression of ~1000 selected genes, representative of the larger
transcriptome.

The goal of this analysis is to

1. Check if we can reliably detect the L1000 genes with DGE
2. Compare drug profiles obtained by DGE with L1000 CMap signatures


* Graph that shows L1000 genes on x-axis, sorted by value on y-axis, y-axis
  is fraction of librares gene was detected in
* Graph comparin L1000 represetnation between CDK and AD data sets. Both axes
  show fraction of libraries gene was detected in
* Functional enrichment comparison DGE vs L1000 for a well known drug treatment
  set, e.g. JAK inhibitors


```{r loading}

comp_res <- syn("syn21559859") %>%
  read_rds()

comp_res_all <- syn("syn21559856") %>%
  read_rds()

gene_meta <- syn("syn21547102") %>%
  read_csv()

signature_meta <- syn("syn21547101") %>%
  read_csv()

counts_raw <- syn("syn21544261") %>%
  read_rds()

dataset_names <- syn("syn21641420") %>%
  read_csv()
```

```{r counts_wrangling}
counts <- counts_raw %>%
  pull(counts) %>%
  bind_rows() %>%
  select(
    ensembl_gene_id, symbol, sample_id, count
  )

meta <- counts_raw %>%
  select(dataset, batch, meta) %>%
  unnest(meta) %>%
  left_join(dataset_names, by = c("dataset", "batch"))

# Using only raw counts for this analysis
l1000_counts <- counts %>%
  mutate(
    l1000 = ifelse(
      symbol %in% (gene_meta %>% filter(pr_is_lm == 1) %>% pull(symbol)),
      "l1000", "not_l1000"
    )
  ) %>%
  inner_join(
    meta,
    by = "sample_id"
  )

```



## L1000 genes in DGE

First, using naive threshold to define "detectable" genes. Requiring at least
5 counts seems adequate.


```{r detectable_l1000}

detectable_threshold <- 5
fraction_detectable <- l1000_counts %>%
  mutate(detectable = count >= detectable_threshold) %>%
  group_by(dataset_name, sample_id, l1000) %>%
  summarize(
    n_detectable = sum(detectable),
    fraction_detectable = n_detectable/n(),
    total_counts = sum(count)
  ) %>%
  ungroup()

fraction_detectable_plot <- fraction_detectable %>%
  ggplot(aes(dataset_name, fraction_detectable, fill = l1000)) +
    geom_violin(draw_quantiles = c(.25, .5, .75)) +
    # facet_grid(vars(method)) +
    labs(
      x = "Dataset",
      y = "Fraction of genes with count â‰¥5",
      title = "Detectable genes per sample",
      fill = "Gene type"
    ) +
    coord_flip()
ggsave(file.path("cmap_overlap", "fraction_detectable.pdf"), fraction_detectable_plot, width = 8, height = 5)

total_counts <- counts %>%
  group_by(sample_id) %>%
  summarize(total = sum(count))

detectable_vs_total_counts <- fraction_detectable %>%
  inner_join(total_counts, by = "sample_id")

detectable_vs_total_counts_plot <- detectable_vs_total_counts %>%
  ggplot(aes(total, fraction_detectable)) +
    geom_point(aes(color = l1000), alpha = 0.5, shape = 16) +
    facet_wrap(vars(dataset_name)) +
    # facet_grid(vars(name), vars(method), scales = "free_x") +
    scale_x_log10() +
    labs(
      title = "Sequencing depth vs fraction of detectable genes",
      x = "Total count per library",
      y = "Fraction of detectable genes",
      color = "Gene type"
    )
ggsave(file.path("cmap_overlap", "sequencing_depth_vs_fraction_detectable.pdf"), detectable_vs_total_counts_plot, width = 9, height = 6)

fraction_detected_per_gene <- l1000_counts %>%
  mutate(detectable = count >= detectable_threshold) %>%
  group_by(dataset_name, ensembl_gene_id, symbol, l1000) %>%
  summarize(n_detectable = sum(detectable), fraction_detectable = n_detectable/n()) %>%
  ungroup()

per_gene_scatter_plot <- function(experiment_1, experiment_2, data, ...) {
  e1 <- sym(experiment_1)
  e2 <- sym(experiment_2)
  # browser()
  ggplot(data, aes(!!e1, !!e2)) +
    geom_point(aes(color = l1000)) +
    labs(title = paste0("Fraction of samples where gene count > 5\n", experiment_1, " vs\n", experiment_2))
}

per_gene_comparison_plots <- combn(unique(fraction_detected_per_gene[["dataset_name"]]), 2) %>%
  t() %>%
  magrittr::set_colnames(c("experiment_1", "experiment_2")) %>%
  as_tibble() %>%
  mutate(
    data = map2(
      experiment_1, experiment_2,
      ~fraction_detected_per_gene %>%
        # Remove bulk, pointless comparison because only cdk dataset has it
        dplyr::filter(dataset_name %in% c(.x, .y)) %>%
        dplyr::select(dataset_name, ensembl_gene_id, l1000, fraction_detectable) %>%
        spread(dataset_name, fraction_detectable, fill = 0)
    )
  ) %>%
  mutate(
    plot = pmap(
      .,
      per_gene_scatter_plot
    )
  )

per_gene_comparison_plots_combined <- per_gene_comparison_plots$plot %>%
  patchwork::wrap_plots(byrow = TRUE, ncol = 1)
ggsave(file.path("cmap_overlap", "fraction_detectable_per_gene_pairwise_comparison.pdf"), per_gene_comparison_plots_combined, width = 5, height = 40)
```

Also plotting how often each gene is detectable across samples.

```{r fraction_detectable_per_gene}
fraction_detectable_per_gene <- l1000_counts %>%
  mutate(detectable = count >= detectable_threshold) %>%
  group_by(comparison, gene_id, l1000) %>%
  summarize(fraction_detectable = sum(detectable)/n()) %>%
  ungroup()

# Output kernel density estimates for each input point defined by vectors x, y
# Adapted from grDevices::densCols to output densities instead of colors
density_2d <- function (x, y = NULL, nbin = 128, bandwidth) {
  xy <- xy.coords(x, y, setLab = FALSE)
  select <- is.finite(xy$x) & is.finite(xy$y)
  x <- cbind(xy$x, xy$y)[select, ]
  map <- grDevices:::.smoothScatterCalcDensity(x, nbin, bandwidth)
  mkBreaks <- function(u) u - diff(range(u))/(length(u) - 1)/2
  xbin <- cut(x[, 1], mkBreaks(map$x1), labels = FALSE)
  ybin <- cut(x[, 2], mkBreaks(map$x2), labels = FALSE)
  dens <- map$fhat[cbind(xbin, ybin)]
  dens[is.na(dens)] <- 0
  dens
}

density_2d_plot <- function (
  data, aesthetics,
  histogram_args = list(bins = 10), scatter_elements = list(), margin_elements = list()
) {
  marginals <- aesthetics[c("x", "y")] %>%
    imap(
      function (axis_var, axis) {
        p <- ggplot(data, aes(!!axis_var)) +
          rlang::exec(
            geom_histogram,
            aes(y = stat(count)/sum(stat(count))),
            !!!histogram_args
          ) +
          geom_step(aes(y = stat(y)), stat = "ecdf") +
          margin_elements
        if (axis == "y") p <- p + scale_y_reverse() + coord_flip()
        p
      }
    )
  density_data <- data %>%
    mutate(density = density_2d(!!aesthetics$x, !!aesthetics$y)) %>%
    arrange(density)
  p <- ggplot(density_data, aesthetics) +
    geom_point() +
    scatter_elements
  egg::ggarrange(
    plots = list(
      ggplot() + theme_void(),
      marginals$x,
      marginals$y,
      p
    ),
    widths = c(1, 4),
    heights = c(1, 3)
  )
}

fraction_detectable_per_gene_density_scatter <- fraction_detectable_per_gene %>%
  spread(comparison, fraction_detectable) %>%
  split(.$l1000) %>%
  imap(
    function(df, gene_set) {
      p <- density_2d_plot(
        df,
        aes(bulk, dge, color = density),
        histogram_args = list(
          breaks = seq(from = 0, to =1, length.out = 10),
          closed = "right"
        ),
        scatter_elements = list(
          coord_fixed(),
          lims(x = c(0, 1), y = c(0, 1)),
          labs(
            x = "RNA-seq (fraction of samples\nwith expression >5)",
            y = "DGE (fraction of samples\nwith expression >5)",
            color = "Density of\ngenes",
            title = gene_set
          ),
          scale_color_viridis_c()
        ),
        margin_elements = list(
          lims(x = c(0, 1)),
          labs(
            x = NULL,
            y = "Fraction\ngenes"
          )
        )
      )
      p
    }
  ) %>%
  wrap_plots(ncol = 1)
ggsave(file.path("cmap_overlap", "fraction_detectable_per_gene_density_scatter.pdf"), fraction_detectable_per_gene_density_scatter, width = 6, height = 8)

```


```{r diff_expr_l1000}
all_ensembl_ids <- comp_res %>%
  pull(result) %>%
  map("ensembl_gene_id") %>%
  reduce(union)

mart <- biomaRt::useMart(
  host = "http://sep2019.archive.ensembl.org",
  biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "hsapiens_gene_ensembl"
)
ensembl_gene_id_mapping_biomart <- biomaRt::select(
  mart, all_ensembl_ids,
  c("hgnc_symbol", "ensembl_gene_id", "gene_biotype"), "ensembl_gene_id"
) %>%
  as_tibble() %>%
  filter(gene_biotype == "protein_coding") %>%
  mutate(hgnc_symbol = magrittr::inset(hgnc_symbol, hgnc_symbol == "", NA_character_)) %>%
  rename(symbol = hgnc_symbol) %>%
  select(-gene_biotype) %>%
  # keep only a single symbol per ensembl_gene_id
  group_by(ensembl_gene_id) %>%
  slice(1) %>%
  ungroup()


l1000_diff_stats <- comp_res_all %>%
  mutate(
    l1000_stats = map(
      result,
      function(r) {
        df <- r %>%
          inner_join(
            ensembl_gene_id_mapping_biomart,
            by = "ensembl_gene_id"
          ) %>%
          inner_join(
            gene_meta %>%
              select(symbol, pr_is_lm),
            by = "symbol"
          )
        df %>%
          mutate(
            de = if_else(padj < 0.05, "de", "nde") %>%
              replace_na("nt") %>%
              factor(levels = c("de", "nde", "nt")),
            lm = if_else(pr_is_lm == 1, "lm", "nlm") %>%
              factor(levels = c("lm", "nlm"))
          ) %>%
          {table(.[["de"]], .[["lm"]])}
        
        # %>%
        #   group_by(de, lm) %>%
        #   summarize(
        #     count = n()
        #   ) %>%
        #   ungroup()
      }
    )
  )

l1000_diff_stats_long <- l1000_diff_stats %>%
  select(-counts_deseq, -deseq, -result) %>%
  mutate(
    frac_lm = map_dbl(l1000_stats, ~.x["de", "lm"] / .x["nde", "lm"]),
    frac_nlm = map_dbl(l1000_stats, ~.x["de", "nlm"] / .x["nde", "nlm"])
  )

ggplot(l1000_diff_stats_long, aes(frac_lm, frac_nlm)) +
  geom_point() +
  coord_equal()

n_diff_exp_scatter <- l1000_diff_stats %>%
  mutate(
    n_lm = map_dbl(l1000_stats, ~.x["de", "lm"]),
    n_nlm = map_dbl(l1000_stats, ~.x["de", "nlm"]),
  ) %>%
  ggplot(aes(n_lm, n_nlm)) +
  geom_point() +
  labs(x = "N differentially expressed L1000", y = "N differentially expressed non-L1000")
ggsave(file.path(wd, "n_diff_exp_scatter.pdf"), n_diff_exp_scatter, width = 4, height = 4)

```


CMap data has been dowloaded from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742


```{r access_cmap_data}
library(webchem)
library(cmapR)

dir.create("cmap_overlap", showWarnings = FALSE)
cmap_paths <- list(
  GSE92742 = "~/data/DGE_comp/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx",
  GSE70138 = "~/data/DGE_comp/GSE70138_Broad_LINCS_Level5_COMPZ_n118050x12328_2017-03-06.gctx"
)

signature_meta <- read_csv(file.path("wrangled", "cmap_signature_meta.csv.gz")) %>%
  mutate(pubchem_cid = as.character(pubchem_cid))

gene_meta <- read_csv(file.path("wrangled", "cmap_gene_meta.csv.gz"))

comp_res_pubchem_ids <- comp_res %>%
  distinct(drug) %>%
  mutate(
    pubchem_id = map(get_cid(str_to_lower(drug), from = "name"), as.character)
  )

comp_res_pubchem_ids_unnested <- comp_res_pubchem_ids %>%
  unnest(pubchem_id) %>%
  drop_na()

# # Fuzzy check if cell lines are called by the same name or slight variations
# cell_line_comp <- stringdist::stringdistmatrix(
#   cdk_meta_combined$cell_line %>% unique() %>% str_to_lower(),
#   signature_meta$cell_id %>% unique() %>% str_to_lower(),
#   useNames = "strings",
#   method = "osa"
# ) %>%
#   as_tibble(rownames = "cdk") %>%
#   gather("cmap", "distance", -cdk) %>%
#   filter(distance <= 2)

shared_conditions <- comp_res %>%
  select(name, method, cell_line, drug, concentration, time, condition) %>%
  group_by(name, method, drug, cell_line) %>%
  arrange(desc(concentration), desc(time), .by_group = TRUE) %>%
  slice(1) %>%
  ungroup() %>%
  mutate(drug = str_to_lower(drug)) %>%
  inner_join(comp_res_pubchem_ids_unnested, by = "drug") %>%
  inner_join(
    signature_meta %>% mutate(cell_id = str_to_lower(cell_id)),
    by = c("pubchem_id" = "pubchem_cid")
  ) %>%
  # Remove controls
  filter(!(pert_iname %in% c("DMSO", "dsmo"))) %>%
  # Remove matches with concentration too different (<50%)
  filter(abs(log10(concentration/pert_dose)) < 0.1760913) %>%
  # Remove matches with time too different (<50%)
  # All CDK data is 24h
  filter(abs(log10(time/pert_time)) < 0.1760913) %>%
  distinct()

# How many unique signatures?
unique_sigs <- shared_conditions %>%
  dplyr::count(sig_id) %>%
  dplyr::count(n)

# Now that we have the IDs of the signatures we can fetch them from the
# big GCTX matrix
cmap_data <- shared_conditions %>%
  group_nest(condition) %>%
  mutate(
    cmap_data = map(
      data,
      function(data) {
        parse.gctx(cmap_paths[[data$dataset[[1]]]], cid = unique(data$sig_id))@mat %>%
          as_tibble(rownames = "pr_gene_id") %>%
          gather("sig_id", "z_score", -pr_gene_id)
      }
    )
  )


cmap_data_prepped <- cmap_data %>%
  ungroup() %>%
  mutate(
    data = map2(data, cmap_data, ~left_join(.y, .x, by = "sig_id"))
  ) %>%
  select(-cmap_data) %>%
  unnest(data) %>%
  left_join(
    gene_meta %>%
      # Convert indicator variables to factor variable
      # showing if gene is directly measured by l1000, computationally inferred
      # with high fidelity (Best INFerred Genes "BING") or inferred with
      # lower confidence
      mutate(
        cmap_measurement = case_when(
          pr_is_lm == 1 ~ "landmark",
          pr_is_bing == 1 ~ "inferred_high_fidelity",
          TRUE ~ "inferred_low_fidelity"
        )
      ) %>%
      select(pr_gene_id, symbol, cmap_measurement) %>%
      mutate(pr_gene_id = as.character(pr_gene_id)),
    by = "pr_gene_id"
  )


```

```{r comparing_deseq_cmap_signatures}
cmap_data_vs_dge <- cmap_data_prepped %>%
  # select(comp_id, sig_id) %>%
  left_join(comp_res %>% unnest(result) %>% select(condition, symbol, log2FoldChange), by = c("condition", "symbol"))
  # select(comp_id, sig_id, cell_line, drug, concentration, time, deseq_res) %>%
  # mutate(deseq_res = map(deseq_res, as_tibble, rownames = "ensembl_gene_id")) %>%
  # unnest(deseq_res) %>%
  # left_join(ensembl_gene_id_mapping, by = "ensembl_gene_id") %>%
  # left_join(cmap_data, by = c("sig_id", "symbol")) %>%
  # drop_na(log2FoldChange, z_score)
```


```{r deseq_cmap_scatter}
deseq_cmap_scatter <- cmap_data_vs_dge %>%
  mutate(cmap_condition = paste(pert_iname, pert_dose, pert_time, cell_id, sep = "\n")) %>%
  group_nest(name, method, cell_line, drug, concentration, time, condition) %>%
  mutate(dge_condition = paste(drug, concentration, time, cell_line, sep = "\n")) %>%
  mutate(
    plot = pmap(
      .,
      function(dge_condition, data, ...) {
          ggplot(data, aes(z_score, log2FoldChange)) +
            # geom_hex() +
            stat_density_2d(
              aes(fill = stat(level)),
              geom = "polygon"
            ) +
            facet_grid(cmap_measurement ~ cmap_condition) +
            scale_fill_viridis_c(trans = "log10") +
            coord_fixed() +
            labs(
              x = "CMap z-score", y = "DGE log2FC",
              fill = "Density",
              title = dge_condition
            )
      }
    )
  )


# ggsave(
#   file.path("cmap_overlap", "cmap_zscore_vs_dge_l2fc_density.pdf"),
#   deseq_cmap_scatter,
#   width = 16, height = 5
# )
# 
# deseq_cmap_cor_test <- deseq_cmap_data %>%
#   mutate(condition = paste(cell_line, drug, concentration, time, sep = "_")) %>%
#   group_by(condition, cmap_measurement) %>%
#   do(
#     test = cor.test(.$z_score, .$log2FoldChange, method = "kendall", conf.int = TRUE) %>% broom::tidy()
#   ) %>%
#   unnest(test) %>%
#   mutate(p.value = sprintf("%.4f", p.value))

```


```{r deseq_cmap_fisher}
def <- tibble(p.value = as.double(NA))
deseq_cmap_fisher <- deseq_cmap_data %>%
  mutate(condition = paste(cell_line, drug, concentration, time, sep = "_")) %>%
  mutate(sig_deseq = padj < 0.05, sig_cmap = abs(z_score) > 1.96) %>%
  group_by(condition, cmap_measurement) %>%
  group_map(
    ~tibble(
      cont_table = list(table(.x[, c("sig_deseq", "sig_cmap")])),
      n_sig_deseq = sum(.x$sig_deseq, na.rm = TRUE),
      n_sig_cmap = sum(.x$sig_cmap, na.rm = TRUE)
    )
  ) %>%
  mutate(
    fisher_test = map(
      cont_table,
      ~possibly(fisher.test, NULL)(.x, conf.int = TRUE, conf.level = .95) %>%
        {if (!is.null(.)) tidy(.) else tibble(p.value = NA)}
    )
  ) %>%
  unnest(fisher_test) %>%
  ungroup()

deseq_cmap_fisher_table <- deseq_cmap_fisher %>%
  arrange(
    group_by(., condition) %>% mutate(p.value = min(p.value, na.rm = TRUE)) %>% pull(p.value),
    condition,
    cmap_measurement
  ) %>%
  select(condition, cmap_measurement, p.value, starts_with("n_sig")) %>%
  mutate(p.value = sprintf("%.03g", p.value)) %>%
  tableGrob(
    theme = ttheme_default(
      base_size = 10
      # core = list(fg_params = list(hjust = 0, x = .05))
    )
  )
ggsave(file.path("cmap_overlap", "cmap_overlap_fisher.pdf"), deseq_cmap_fisher_table, width = 7, height = 10)
```



