---
title: "CDK raw files"
author: "Clemens Hug"
date: "2/23/2019"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(here)
library(synExtra)
synapser::synLogin()
syn <- synExtra::synDownloader("data")

```

## Locate raw CDK4/6 files on o2

Files are located on `/n/files/ImStor` mount, have to use the transfer node to
access them.

Mount the file system using `sshfs` tool.

```{bash mount_imstor}
mkdir -p ImStor
sshfs ch305@transfer.rc.hms.harvard.edu:/n/files/ImStor ImStor
```

Fetch list of fastq files.

```{r fast_files}
fastqs_list <- list(
  "201608" = list.files(
    file.path(
      "ImStor/sorger/data/rnaseq/lincs/rnaseq/CDK46_response_201608/fastq",
      c("160824_CM3415-1_NS500233_fastq", "160825_CM3415-2_NS500144_fastq")
    ),
    pattern = ".+\\.fastq\\.gz",
    full.names = TRUE
  ),
  "201701" = list.files(
    "ImStor/sorger/data/rnaseq/lincs/rnaseq/CDK46_response_201701/fastq",
    pattern = ".+\\.fastq\\.gz",
    full.names = TRUE
  )
) %>%
  map(~paste("/n/files", .x, sep = "/"))
```


```{r metadata}
metadata_list <- list(
  "201608" = read_csv("ImStor/sorger/data/rnaseq/lincs/rnaseq/CDK46_response_201608/CDK46_response_201608_metadata.csv"),
  "201701" = read_csv("ImStor/sorger/data/rnaseq/lincs/rnaseq/CDK46_response_201701/CDK46_response_201701_metadata.csv")
)

```

```{bash unmount_imstor}
diskutil unmount ImStor
rmdir ImStor
```


```{r combine_data}
fastqs <- fastqs_list %>%
  map(enframe, value = "fastq") %>%
  bind_rows(.id = "run")

metadata <- list(
  "201608" = metadata_list[["201608"]] %>%
    transmute(
      sample_id = paste("201608", Well, sep = "_"),
      well = Well,
      cells = `cell line`,
      drug,
      drug_conc = dose,
      time = time %>%
        str_replace(fixed("h"), "") %>%
        as.integer()
    ) %>%
    drop_na(sample_id),
  "201701" = metadata_list[["201701"]] %>%
    transmute(
      sample_id = paste("201701", Sample, sep = "_"),
      well = Position,
      # According to Kartik all samples in this experiment (201701) are MCF7
      cells = "MCF7",
      Treatment,
      time = `Time point` %>%
        str_replace(fixed(" hours"), "") %>%
        as.integer()
    ) %>%
    separate(Treatment, c("drug_conc", "drug"), sep = " ", fill = "left", extra = "merge") %>%
    mutate(drug_conc = as.numeric(replace_na(drug_conc, 0)))
) %>%
  bind_rows(.id = "run") %>%
  mutate(
    drug = recode(
      drug, Palbo = "palbociclib", Abema = "abemaciclib", Ribo = "ribociclib",
      ctrl = "control", dmso = "control"
    )
  ) %>%
  mutate(
    condition = do.call(
      paste,
      as.list(.)[c("cells", "stim", "stim_conc", "drug", "drug_conc", "time")] %>%
        magrittr::inset2("sep", "_")
    ),
    control_condition = do.call(
      paste,
      as.list(.)[c("cells", "stim", "stim_conc", "drug", "drug_conc", "time")] %>%
        magrittr::inset2("drug", "control") %>%
        magrittr::inset2("drug_conc", 0) %>%
        magrittr::inset2("sep", "_")
    ) %>%
      magrittr::inset(!(. %in% condition), NA_character_)
  ) %>%
  # Get rid of irrelevant samples
  drop_na(time)

write_csv(
  metadata, here("wrangled", "meta_cdk4_6_bulk.csv")
)

metadata_fastq <- fastqs %>%
  mutate(
    sample_id = paste(run, str_match(basename(fastq), "^[^_]+_([^_]+)_.+")[,2], sep = "_")
  ) %>%
  inner_join(metadata, by = c("run", "sample_id"))

write_csv(
  metadata_fastq, here("wrangled", "cdk_bulk_fastq.csv")
)

```

## ERCC spike-in

The bulk RNA-seq samples where run with ERCC spike in, adding the fasta files
for alignment from https://assets.thermofisher.com/TFS-Assets/LSG/manuals/ERCC92.zip

Move to 

```{r ercc}
download.file(
  "https://assets.thermofisher.com/TFS-Assets/LSG/manuals/ERCC92.zip",
  file.path("..", "wrangled", "ercc_fasta.zip")
)
```

```{bash unpack_ercc}
unzip -o -d ../wrangled ../wrangled/ercc_fasta.zip
ssh ch305@o2.hms.harvard.edu 'mkdir -p /home/ch305/annotation/ercc/'
scp ../wrangled/ERCC92.fa ch305@o2.hms.harvard.edu:/home/ch305/annotation/ercc/
```

## Bulk metadata for bcbio

Format cdk data for alignment with Bcbio using the o2jobber utility (on github
at https://github.com/clemenshug/o2jobber )


```{r metadata_alignment}
metadata_alignment <- metadata_fastq %>%
  transmute(
    id = sample_id,
    fastq = paste0("transfer.rc.hms.harvard.edu:", fastq),
    transcriptome_fasta = "/home/ch305/annotation/hsap/GRCh38.98/Homo_sapiens.GRCh38.combined.fa",
    transcriptome_gtf = "/home/ch305/annotation/hsap/GRCh38.98/Homo_sapiens.GRCh38.98.gtf",
    spikein_fasta = "/home/ch305/annotation/ercc/ERCC92.fa"
  )
write_csv(
  metadata_alignment,
  here("wrangled", "cdk_bulk_alignment_metadata.csv")
)
```

Followed by using o2jobber to start alignment job with bcbio on O2.

```{r synapse}
activity <- synapser::Activity(
  name = "Wrangle bulk CDK4/6 inhibitor data",
  executed = "https://github.com/clemenshug/dge_comparison/blob/master/wrangling/01_cdk_raw_files.Rmd"
)

c(
  here("wrangled", "meta_cdk4_6_bulk.csv")
) %>%
  synStoreMany(parentId = "syn21541011", activity = activity)

#rsync -rv --exclude *.fq.gz --exclude fastq --exclude *.fastq.gz --exclude *.bam --exclude rapmap --exclude transcriptome 2018_02_steven_rodriguez_ad_repurposing transfer:/n/files/ImStor/sorger/data/rnaseq/lincs/dge_reanalysis_clemens/
```

