---
title: "CMap data"
author: "Clemens Hug"
date: "3/19/2019"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(cmapR)
library(webchem)
library(synExtra)
library(here)
library(lspcheminf)
library(furrr)

synapser::synLogin()
syn <- synExtra::synDownloader("~/data/DGE_comp/")
```


```{r compound_info}
compound_dict <- syn("syn20835543") %>%
  read_rds()

fingerprints <- syn("syn21042105") %>%
  read_rds()

# Using previously done pertubation meta and signature meta
# pm <- syn("syn21547097") %>%
#   read_csv()
# 
# im <- syn("syn21547100") %>%
#   read_csv()
# 
# sm <- syn("syn21547101") %>%
#   read_csv()

```


## CMap data

There are two big CMap datasets, Phase I and II with two different GEO accession
numbers. Loading metadata and Z-score matrix from both.


```{r perturbation_meta}
pertubation_meta_raw <- list(
  GSE92742 = "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE92nnn/GSE92742/suppl/GSE92742_Broad_LINCS_pert_info.txt.gz",
  GSE70138 = "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE70nnn/GSE70138/suppl/GSE70138_Broad_LINCS_pert_info_2017-03-06.txt.gz"
) %>%
  map(read_tsv, na = "-666") %>%
  bind_rows(.id = "dataset")

unique_compounds <- pertubation_meta_raw %>%
  drop_na(canonical_smiles) %>%
  distinct(pert_id, smiles = canonical_smiles) %>%
  left_join(
    convert_compound_identifier(
      set_names(.$smiles, .$pert_id),
      identifier = "smiles",
      target_identifier = "inchi"
    ) %>%
      select(inchi = compounds, pert_id = names),
    by = "pert_id"
  )



plan(multisession(workers = 8))
unique_compounds_canonical <- unique_compounds %>%
  distinct(pert_id, inchi) %>%
  drop_na() %>%
  chunk_df(16) %>%
  future_map(
    ~left_join(
      .x,
      canonicalize_compound(
        set_names(.x$inchi, .x$pert_id),
        standardize = TRUE
      ) %>%
        select(pert_id = compound, canonical_inchi = inchi),
      by = "pert_id"
    ),
    .progress = TRUE
  ) %>%
  bind_rows()

write_rds(
  unique_compounds_canonical,
  here("wrangled", "cmap_compounds_canonical.rds"),
  compress = "gz"
)

# unique_compounds_canonical <- read_rds(
#   here("wrangled", "cmap_compounds_canonical.rds")
# )


unique_compounds_fingerprints <- unique_compounds_canonical %>%
  drop_na(canonical_inchi) %>%
  crossing(
    fingerprint_type = c("morgan", "topological")
  ) %>%
  group_nest(fingerprint_type) %>%
  mutate(
    data = map2(
      data, fingerprint_type,
      function(df, fpt) {
        q <- set_names(df[["canonical_inchi"]], df[["pert_id"]])
        attr(q, "identifier") <- "inchi"
        lspcheminf::calculate_fingerprints(q, fpt)
      }
    )
  )

write_rds(
  unique_compounds_fingerprints,
  here("wrangled", "cmap_compound_fingerprints.rds"),
  compress = "gz"
)

# unique_compounds_fingerprints <- read_rds(
#   here("wrangled", "cmap_compound_fingerprints.rds")
# )

lsp_fp <- fingerprints %>%
  filter(fp_name == "morgan_normal") %>%
  chuck("data", 1) %>%
  filter(fp_name != "morgan_chiral") %>%
  group_nest(fingerprint_type = fp_type, .key = "reference")

compound_lspci_id_mapping <- unique_compounds_fingerprints %>%
  left_join(
    lsp_fp, by = "fingerprint_type"
  ) %>%
  mutate(
    data = map2(
      data, reference,
      function(d, r) {
        dq <- set_names(d[["fingerprints"]], d[["names"]])
        rq <- set_names(r[["fingerprint"]], as.character(r[["lspci_id"]]))
        # browser()
        lspcheminf::chemical_similarity_threshold(
          dq, rq, precalculated = TRUE, threshold = 0.99999,
          n_threads = 10
        )
      }
    )
  )

compound_dict_chunks <- compound_dict %>%
  filter(fp_name == "morgan_normal") %>%
  chuck("data", 1) %>%
  drop_na(inchi) %>%
  select(lspci_id, inchi) %>%
  chunk_df(16)

compound_dict_mw <- compound_dict_chunks %>%
  future_map(
    ~lspcheminf::molecular_mass(
      set_names(.x$inchi, .x$lspci_id),
      url = "http://127.0.0.1:8001"
    ),
    .progress = TRUE
  )

compound_dict_mw_df <- compound_dict_mw %>%
  bind_rows() %>%
  transmute(
    lspci_id = as.integer(compound),
    mass
  )

write_rds(
  compound_dict_mw_df,
  here("wrangled", "compound_dict_mw.rds"),
  compress = "gz"
)

cmap_mw <- unique_compounds_canonical %>%
  drop_na(canonical_inchi) %>%
  {set_names(.[["canonical_inchi"]], .[["pert_id"]])} %>%
  lspcheminf::molecular_mass(url = "http://127.0.0.1:8001")

compound_lspci_id_mapping_df <- compound_lspci_id_mapping %>%
  select(data, fingerprint_type) %>%
  unnest(data) %>%
  mutate_at(vars(target), as.integer) %>%
  left_join(
    select(compound_dict_mw_df, lspci_id, mass_lspci_id = mass),
    by = c("target" = "lspci_id")
  ) %>%
  left_join(
    select(cmap_mw, pert_id = compound, mass_pert_id = mass),
    by = c("query" = "pert_id")
  ) %>%
  # Remove matches with unqual mass
  filter(near(mass_lspci_id, mass_pert_id)) %>%
  group_by(query) %>%
  filter(
    n() == 2,
    all(c("morgan", "topological") %in% fingerprint_type),
    length(unique(target)) == 1
  ) %>%
  ungroup() %>%
  distinct(pert_id = query, lspci_id = target)


write_csv(
  compound_lspci_id_mapping_df,
  here("wrangled", "cmap_pertubation_compound_mapping.csv")
)

pertubation_meta <- pertubation_meta_raw %>%
  mutate(
    pert_iname_norm = str_replace_all(pert_iname, "[^a-zA-Z\\d]", "") %>% str_to_lower()
  ) %>%
  left_join(
    compound_lspci_id_mapping_df,
    by = "pert_id"
  )
  # left_join(pubchem_ids, by = "inchi_key") %>%
  # # Some compounds in the original dataset already have pubchem ids, merging
  # # them with the fetched ones
  # mutate(
  #   pubchem_cid = map2(
  #     pubchem_cid.x, pubchem_cid.y,
  #     ~unique(c(.x, .y)) %>%
  #       # Remove NA values if there is a non-NA value
  #       {if (sum(!is.na(.)) > 0) na.omit(.) else .}
  #   )
  # ) %>%
  # select(-pubchem_cid.x, -pubchem_cid.y) %>%
  # unnest(pubchem_cid)

# pertubation_meta %>% filter(!is.na(lspci_id))

write_csv(pertubation_meta, here("wrangled", "cmap_perturbation_meta.csv.gz"))
```



```{r instance_meta}
instance_meta_raw <- list(
  GSE92742 = "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE92nnn/GSE92742/suppl/GSE92742_Broad_LINCS_inst_info.txt.gz",
  GSE70138 = "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE70nnn/GSE70138/suppl/GSE70138_Broad_LINCS_inst_info_2017-03-06.txt.gz"
) %>%
  map(read_tsv, na = "-666") %>%
  bind_rows(.id = "dataset")

instance_meta <- instance_meta_raw %>%
  left_join(distinct(pertubation_meta, pert_id, pubchem_cid, lspci_id), by = "pert_id")

write_csv(instance_meta, file.path("..", "wrangled", "cmap_instance_meta.csv.gz"))
```




```{r signature_meta}
signature_meta_raw <- list(
  GSE92742 = "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE92nnn/GSE92742/suppl/GSE92742_Broad_LINCS_sig_info.txt.gz",
  GSE70138 = "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE70nnn/GSE70138/suppl/GSE70138_Broad_LINCS_sig_info_2017-03-06.txt.gz"
) %>%
  map(read_tsv, na = c("-666", "-666.00")) %>%
  bind_rows(.id = "dataset")
# Parsing failures correspond to malformed fields in input data, are in irrelevant
# signatures, so ignore...

signature_meta <- signature_meta_raw %>%
  mutate(
    pert_dose = if_else(is.na(pert_dose), as.double(str_match(pert_idose, "([\\d\\.]+) um")[, 2]), pert_dose) %>%
      {if_else(. < 0, NA_real_, .)},
    pert_dose_unit = if_else(str_detect(pert_idose, "([\\d\\.]+) um"), "ÂµM", pert_dose_unit),
    pert_time = if_else(is.na(pert_time), as.double(str_match(pert_itime, "([\\d]+) h")[, 2]), pert_time),
    pert_time_unit = if_else(is.na(pert_itime), NA_character_, "h")
  ) %>%
  left_join(distinct(pertubation_meta, pert_id, lspci_id), by = "pert_id")

write_csv(signature_meta, file.path("..", "wrangled", "cmap_signature_meta.csv.gz"))
```


```{r gene_meta}
gene_meta_raw <- list(
  GSE92742 = "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE92nnn/GSE92742/suppl/GSE92742_Broad_LINCS_gene_info.txt.gz",
  GSE70138 = "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE70nnn/GSE70138/suppl/GSE70138_Broad_LINCS_gene_info_2017-03-06.txt.gz"
) %>%
  map(read_tsv, na = "-666") %>%
  bind_rows(.id = "dataset")

# Check if the two datasets are identical
gene_meta_identical <- gene_meta_raw %>%
  group_by_at(vars(-dataset)) %>%
  summarize(n = n()) %>%
  ungroup() %>%
  dplyr::count(n)
# They are!

gene_meta <- gene_meta_raw %>%
  filter(dataset == "GSE92742") %>%
  mutate_at(vars(pr_gene_id), as.character) %>%
  genebabel::join_hgnc(
    "pr_gene_id",
    "entrez_id",
    c("symbol", "entrez_id", "ensembl_gene_id")
  ) %>%
  select(-dataset)

write_csv(gene_meta, file.path("..", "wrangled", "cmap_gene_meta.csv.gz"))
```

Gene metadata for both datasets are identical, can save just a single one.

```{r synapse}
activity <- synapser::Activity(
  name = "Wrangle CMap data",
  used = c(
    "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE92nnn/GSE92742/suppl/GSE92742_Broad_LINCS_pert_info.txt.gz",
    "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE70nnn/GSE70138/suppl/GSE70138_Broad_LINCS_pert_info_2017-03-06.txt.gz",
    "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE92nnn/GSE92742/suppl/GSE92742_Broad_LINCS_inst_info.txt.gz",
    "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE70nnn/GSE70138/suppl/GSE70138_Broad_LINCS_inst_info_2017-03-06.txt.gz",
    "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE92nnn/GSE92742/suppl/GSE92742_Broad_LINCS_sig_info.txt.gz",
    "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE70nnn/GSE70138/suppl/GSE70138_Broad_LINCS_sig_info_2017-03-06.txt.gz",
    "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE92nnn/GSE92742/suppl/GSE92742_Broad_LINCS_gene_info.txt.gz",
    "ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE70nnn/GSE70138/suppl/GSE70138_Broad_LINCS_gene_info_2017-03-06.txt.gz"
  ),
  executed = "https://github.com/clemenshug/dge_comparison/blob/master/wrangling/05_cmap_data.Rmd"
)

c(
  here("wrangled", "cmap_perturbation_meta.csv.gz"),
  here("wrangled", "cmap_instance_meta.csv.gz"),
  here("wrangled", "cmap_signature_meta.csv.gz"),
  here("wrangled", "cmap_gene_meta.csv.gz")
) %>%
  synStoreMany(parentId = "syn21547022", activity = activity)

#rsync -rv --exclude *.fq.gz --exclude fastq --exclude *.fastq.gz --exclude *.bam --exclude rapmap --exclude transcriptome 2018_02_steven_rodriguez_ad_repurposing transfer:/n/files/ImStor/sorger/data/rnaseq/lincs/dge_reanalysis_clemens/
```


