
```{r setup}
library(tidyverse)
library(synExtra)
library(readxl)
library(here)

synapser::synLogin()
syn <- synExtra::synDownloader("~/data", .cache = TRUE)

wd <- here("wrangled")
```


```{r wrangling_funcs}
barcode_well_map <- syn("syn12979100") %>%
  read_xlsx(sheet = "barcodes_trugrade_384_set1")
  # rename(Barcode = barcode, Well = well, Plate_ID = plate_id)

prepare_barcode_counts <- function(barcode_count_file, barcode_map) {
  read_tsv(barcode_count_file, col_names = c("barcode", "count")) %>%
    mutate(
      barcode = str_replace(barcode, "^.*([ATCG]{6})$", "\\1")
    ) %>%
    left_join(
      barcode_well_map, by = "barcode"
    )
}

prepare_col_meta <- function(colmeta_file, barcode_map) {
  read_csv(colmeta_file, col_names = "barcode") %>%
    mutate(
      barcode = if (all(str_detect(barcode, fixed(":")))) str_split(barcode, fixed(":"), simplify = TRUE)[, 2] else barcode,
      col_idx = 1:n()
    ) %>%
    left_join(
      barcode_map %>%
        select(barcode, well),
      by = "barcode"
    )
}

prepare_row_meta <- function(rowmeta_file) {
  read_csv(rowmeta_file, col_names = "ensembl_gene_id") %>%
    mutate(row_idx = 1:n())
}

prepare_mtx_sparse <- function(mtx_file) {
  read_delim(
    mtx_file,
    delim = " ",
    skip = 3,
    col_names = c("row_idx", "col_idx", "count"),
    col_types = "iii"
  )
}

prepare_counts <- function(mtx, col_meta, row_meta, gene_mapping) {
  # browser()
  mtx %>%
    left_join(col_meta, by = "col_idx") %>%
    left_join(row_meta, by = "row_idx") %>%
    select(well, ensembl_gene_id, count) %>%
    inner_join(
      gene_mapping, by = "ensembl_gene_id"
    )
}

norm_drug <- function(x) {
  x %>%
    str_to_lower() %>%
    str_replace_all("[^a-zA-Z0-9]", "")
}

```

## John Santa Maria Merck drugs

Load and process RNA-seq data from John Santa Maria

John Santa Maria initiated a project looking at gene expression underlying phenotypic changes in intestinal myofibroblasts.

Cells were looked at for one time point at 3 drug doses in triplicate.  This was done either in the presence or absense of activating stimulus IL-1B.


-These cells were purchased from Lonza and originate from terminal fetuses. They are intestinal myofibroblasts.
-These are immunomodulatory drugs, most of which are public, save 2 which are private Merck compounds. All the data is jointly owned by Sorger and Merck, but the 2 private structures may be retained by the company (they also may not, as they are not of general interest).

- Sarah is in touch with Jon to get more details on the treatment of the cells.

Update 7_25_18

Primary human intestinal myofibroblasts were treated with a collection of tool compounds in dose series. The fibroblasts were treated with incubated with drugs for 24 hours. Then, either vehicle or 0.1 ng/mL IL-1B for 0, 1, or 6 hours before harvesting for DGE-Seq.


transfer.rc.hms.harvard.edu:/n/files/ImStor/sorger/data/rnaseq/lincs/DGE_fastq_files/cdk46_runs_batch2/170831_NS500422_0543_AHWKJLBGX2/

## Metadata

```{r jsm_meta}

jsm_meta_raw <- syn("syn21541012") %>%
  read_tsv()

jsm_meta <- jsm_meta_raw %>%
  transmute(
    plate = 1,
    replicate = 1,
    well = str_replace(Well, "^([A-Z])([1-9])$", "\\10\\2"),
    cells = Cells,
    stim = recode(Stim, "0" = "control"),
    stim_conc = `StimConc(inng/mL)`,
    # Misspelled drug
    drug = recode(
      Drug,
      "0" = "control",
      "5Z-7-Oxozeanol" = "5Z-7-Oxozeaenol",
      "Nec-1s" = "CHEMBL381230"
    ),
    drug_conc = `DrugConc(inuM)`
  )

```


```{r lincs_cdk4_6_old_meta}
cdk_old_meta_raw <- synapser::synGet("syn21542403", followLink = TRUE) %>%
  chuck("path") %>%
  read_csv()

cdk_old_meta <- cdk_old_meta_raw %>%
  transmute(
    plate = 1,
    replicate = 1,
    well,
    cells = cell_line,
    drug = recode(drug, "DMSO" = "control"),
    drug_conc = concentration__um,
    time = 6
  )

```


```{r lincs_cdk4_6_7_new_meta}
cdk_new_meta_raw <- syn("syn21544269") %>%
  read_csv()

cdk_new_meta <- cdk_new_meta_raw %>%
  drop_na(cell_line) %>%
  transmute(
    plate,
    replicate = plate,
    well,
    cells = cell_line,
    drug  = recode(agent, "DMSO" = "control"),
    drug_conc = concentration,
    time = timepoint
  )

```


```{r ld_dub_meta}
ld_dub_raw <- synapser::synGet("syn21542711", followLink = TRUE) %>%
  chuck("path") %>%
  read_tsv()

ld_dub_meta <- ld_dub_raw %>%
  transmute(
    plate = 1,
    replicate = 1,
    well,
    cells = cellline,
    drug = recode(
      agent,
      "DMSO" = "control",
      "CGM097" = "CHEMBL3601398"
    ),
    drug_conc = concentration,
    time
  )

```


```{r sr_repurposing_meta}
sr_repurposing_meta_raw1 <- synapser::synGet("syn21546999", followLink = TRUE) %>%
  chuck("path") %>%
  read_csv()

sr_repurposing_meta_raw2 <- synapser::synGet("syn21547000", followLink = TRUE) %>%
  chuck("path") %>%
  read_csv()

sr_repurposing_meta1 <- sr_repurposing_meta_raw1 %>%
  transmute(
    plate = 1,
    replicate = 1,
    well = Well,
    cells = "rencell",
    drug = recode(Drug, "Drug control" = "control"),
    drug_conc = Concentration %>%
      magrittr::inset(is.na(.), 0)
  )

sr_repurposing_meta2 <- sr_repurposing_meta_raw2 %>%
  transmute(
    plate = 1,
    replicate = 1,
    well = Well,
    cells = "rencell",
    drug = recode(Drug, "DMSO" = "control"),
    stim = str_match(drug, "\\((.+)\\)$") %>%
      {.[, 2]} %>%
      str_trim(),
    drug_conc = Concentration %>%
      magrittr::inset(is.na(.), 0)
  ) %>%
  mutate(
    stim = if_else(drug %in% stim, drug, stim),
    drug = str_replace_all(drug, "\\(.+\\)$", "") %>%
      str_trim() %>%
      # If drug is actually a stimulus put control
      {if_else(. == stim & !is.na(stim), "control", .)}
  )

```



```{r fp_transdiff_screen_meta}
fp_transdiff_meta_raw <- synapser::synGet("syn22005605", followLink = TRUE) %>%
  chuck("path") %>%
  read_csv()

fp_transdiff_meta <- fp_transdiff_meta_raw %>%
  gather("column", "drug", -row) %>%
  transmute(
    # Add leading zero to column number
    well = paste0(row, sprintf("%02d", as.integer(column))),
    drug = recode(drug, DMSO = "control"),
    cells = "mouse_embryonic_fibroblasts",
    # Exact concentration unknown...
    drug_conc = if_else(drug != "control", 1, 0),
    replicate = 1,
    time = 24
  ) %>%
  # Experiment contains 4 identical plates
  crossing(plate = 1:4)
```


```{r optimal_kinase_library_meta}
lspci_id_canonical_names <- syn("syn26230061") %>%
  read_csv()

lspci_id_vendor_id_map <- syn("syn26230064") %>%
  read_csv()

okl_meta_raw <- syn("syn25764541") %>%
  read_csv()

okl_meta <- okl_meta_raw %>%
  left_join(
    lspci_id_vendor_id_map,
    by = c("agent" = "vendor_id")
  ) %>%
  left_join(
    lspci_id_canonical_names,
    by = c("lspci_id")
  ) %>%
  transmute(
    plate = 1,
    replicate = 1,
    well,
    cells = cell_line,
    drug = if_else(agent == "DMSO", "control", name),
    lspci_id,
    drug_conc = concentration,
    time = timepoint
  )

```

```{r mixed_dge_23-02-10_meta}
mixed_meta_raw <- syn("syn51181405") %>%
  read_csv()

mixed_meta <- mixed_meta_raw %>%
  drop_na(agent) %>%
  transmute(
    plate = source_96,
    well = well_384,
    cells = cell_line,
    drug = if_else(agent == "DMSO", "control", agent),
    drug_conc = concentration,
    time = timepoint
  ) %>%
  replace_na(list(time = 24)) %>%
  group_by(cells, drug, drug_conc, time) %>%
  mutate(
    replicate = seq_len(n())
  ) %>%
  ungroup()
```



```{r albers_dge_2023-04_meta}
albers_meta_raw <- syn("syn51716406") %>%
  read_csv()

albers_meta <- albers_meta_raw %>%
  drop_na(drug) %>%
  transmute(
    plate = source_96,
    well = well_384,
    cells = "rencell",
    drug = if_else(drug == "DMSO", "control", drug),
    drug_conc = concentration_uM,
    time = 24
  ) %>%
  group_by(cells, drug, drug_conc, time) %>%
  mutate(
    replicate = seq_len(n())
  ) %>%
  ungroup()
```


```{r counts}
dataset_syn <- tribble(
  ~dataset, ~folder_syn, ~date,
  "jsm_merck", "syn21541313", "2017_09",
  "lincs_cdk4_6_7", "syn21541462", "2017_09",
  "ld_dub", "syn21541471", "2018_06",
  "lincs_cdk4_6_7", "syn21541479", "2019_08",
  "sr_repurposing", "syn21548149", "2018_02",
  "sr_repurposing", "syn21548150", "2018_11",
  "fp_transdiff", "syn22041289", "2015_10",
  "okl", "syn26156493", "2021_08",
  "mixed_dge", "syn51181384", "2023_02",
  "albers_dge", "syn51716389", "2023_04"
) %>%
  mutate(
    files = map(
      folder_syn,
      function(x) {
        folder_content = synChildren(x)
        cm_match <- str_match(names(folder_content), "^(?:(.*([0-9]+))|tagcounts)\\.mtx$")
        cm_hit <- !is.na(cm_match[, 1])
        # browser()
        if (sum(cm_hit) > 1) {
          plate_idx <- cm_match[, 3][cm_hit]
          file_prefix <- cm_match[, 2][cm_hit]
          tibble(
            plate = plate_idx,
            counts = folder_content[cm_hit],
            count_dupes = folder_content[paste0(file_prefix, "-dupes.mtx")],
            colnames = folder_content[paste0(file_prefix, ".mtx.colnames")],
            rownames = folder_content[paste0(file_prefix, ".mtx.rownames")],
            barcode_counts = folder_content[paste0(file_prefix, "-barcodes.tsv")]
          )
        } else {
          tibble(
            plate = "1",
            counts = folder_content["tagcounts.mtx"],
            count_dupes = folder_content["tagcounts-dupes.mtx"],
            colnames = folder_content["tagcounts.mtx.colnames"],
            rownames = folder_content["tagcounts.mtx.rownames"],
            barcode_counts = folder_content["cb-histogram.txt"]
          )
        }
      }
    )
  ) %>%
  unnest(files)

dataset_files <- dataset_syn %>%
  mutate_at(vars(-dataset, -folder_syn, -date), map_chr, syn)

datasets_raw <- dataset_files %>%
  mutate_at(
    vars(counts, count_dupes),
    map, prepare_mtx_sparse
  ) %>%
  mutate(
    barcode_counts = map(barcode_counts, prepare_barcode_counts, barcode_well_map) %>%
      map(
        function(x) {
          # Only keep P11 well with highest count
          # browser()
          x %>%
            mutate(well_id = str_split_fixed(well, fixed("_"), 2)[, 1]) %>%
            arrange(well_id, desc(count)) %>%
            group_by(well_id) %>%
            slice(1) %>%
            ungroup() %>%
            select(barcode, well = well_id, count)
        }
      ),
    barcode_map = map(
      barcode_counts,
      select,
      well, barcode
    ),
    colnames = map2(colnames, barcode_map, prepare_col_meta),
    rownames = map(rownames, prepare_row_meta)
  )

if(!file.exists("~/data/genenames_ensembl.txt"))
  download.file(
    "https://www.genenames.org/cgi-bin/download/custom?col=gd_hgnc_id&col=gd_app_sym&col=gd_app_name&col=gd_status&col=gd_prev_sym&col=gd_aliases&col=gd_pub_chrom_map&col=gd_pub_acc_ids&col=gd_pub_refseq_ids&col=gd_locus_type&col=gd_locus_group&col=gd_pub_ensembl_id&col=md_ensembl_id&status=Approved&status=Entry%20Withdrawn&hgnc_dbtag=on&order_by=gd_app_sym_sort&format=text&submit=submit",
    "~/data/genenames_ensembl.txt",
    method = "curl"
  )

genenames_raw <- read_tsv(
  "~/data/genenames_ensembl.txt"
)

genenames_raw %>%
  distinct(`Ensembl gene ID`)

all_ensembl_ids <- datasets_raw %>%
  pull(rownames) %>%
  map("ensembl_gene_id") %>%
  reduce(union)

setdiff(all_ensembl_ids, genenames_raw$`Ensembl gene ID`) %>%
  length()
length(all_ensembl_ids)

mart <- biomaRt::useMart(
  # host = "http://ensembl.org",
  biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "hsapiens_gene_ensembl"
)
ensembl_gene_id_mapping_biomart <- biomaRt::select(
  mart, all_ensembl_ids,
  c("hgnc_symbol", "ensembl_gene_id", "gene_biotype"), "ensembl_gene_id"
) %>%
  as_tibble() %>%
  filter(gene_biotype == "protein_coding") %>%
  # Replace empty symbols with NA
  mutate(hgnc_symbol = magrittr::inset(hgnc_symbol, hgnc_symbol == "", NA_character_)) %>%
  rename(symbol = hgnc_symbol) %>%
  select(-gene_biotype) %>%
  # keep only a single symbol per ensembl_gene_id
  group_by(ensembl_gene_id) %>%
  slice(1) %>%
  ungroup()

# There are some spurios barcodes that don't map to any well
# get rid of them with drop_na
datasets_counts <- datasets_raw %>%
  transmute(
    dataset,
    plate,
    date,
    counts = pmap(
      list(counts, colnames, rownames),
      prepare_counts,
      ensembl_gene_id_mapping_biomart
    ) %>%
      map(drop_na, well),
    count_dupes = pmap(
      list(count_dupes, colnames, rownames),
      prepare_counts,
      ensembl_gene_id_mapping_biomart
    ) %>%
      map(drop_na, well),
    barcode_counts = barcode_counts %>%
      map(drop_na, well)
  )

```


```{r datasets_with_meta}
meta_all <- tribble(
  ~dataset, ~meta, ~date,
  "jsm_merck", jsm_meta, "2017_09",
  "lincs_cdk4_6_7", cdk_old_meta, "2017_09",
  "ld_dub", ld_dub_meta, "2018_06",
  "lincs_cdk4_6_7", cdk_new_meta, "2019_08",
  "sr_repurposing", sr_repurposing_meta1, "2018_02",
  "sr_repurposing", sr_repurposing_meta2, "2018_11",
  "fp_transdiff", fp_transdiff_meta, "2015_10",
  "okl", okl_meta, "2021_08",
  "mixed_dge", mixed_meta, "2023_02",
  "albers_dge", albers_meta, "2023_04"
) %>%
  mutate(
    across(meta, map, ~mutate(.x, across(plate, as.character)))
  ) %>%
  unnest(meta) %>%
  mutate(
    drug_norm = norm_drug(drug),
    plate = if_else(is.na(plate), "1", as.character(plate)),
    batch = paste(dataset, date, plate, sep = "_"),
  )

# drug_name_mapping <- syn("syn26227670") %>%
#   read_csv()

drug_name_mapping <-  syn("syn22035396") %>%
  read_rds() %>%
  filter(fp_name == "morgan_normal") %>%
  chuck("data", 1)

drug_name_mapping_norm <- drug_name_mapping %>%
  transmute(
    lspci_id,
    name,
    name_norm = norm_drug(name),
    source_collapsed
  )

  # Barasertib appears twice in Chembl (CHEMBL415049 and CHEMBL215152)
  # CHEMBL215152 is most likely a mistake
  # keeping only CHEMBL415049 (lspci_id 102882), discarding CHEMBL215152 (lspci_id 100354)
  # Same thing with IMD-0453 and omeprazole
  # Pifithrin-mu in this case not what was used, pifithrin-alpha
  # Piceantannol has weird secondary naming in 90990, primary is in 2676
  # NPS 2143 is misannotated at one vendor
  # Candesartan was used, not Candesartan cilexetil
  # AP26113 misannotated at chembl with alt name
  # Tolfenamic acid misannotated at vendor
  # AT101 misannoated at ChEMBL with alt name
  # NLG919 misannotated at vendor
  # PYR41 misannoated at vendor
  # filter(
  #   !lspci_id %in% c(100354, 1443860, 149325, 634201, 80990, 392021, 87456, 1444031, 1100176, 107269, 115787, 1207732)
  # )

cmdps_mapped <- meta_all %>%
  distinct(
    drug,
    drug_norm
  ) %>%
  mutate(
    drug_all = map(
      drug_norm,
      ~str_match(.x, "\\s\\((.+)\\)$")[, 2] %>%
        str_split(",\\s?") %>%
        c(.x) %>%
        unlist() %>%
        na.omit()
    )
  ) %>%
  unchop(drug_all) %>%
  distinct() %>%
  left_join(
    distinct(drug_name_mapping_norm, lspci_id, name, name_norm, source_collapsed),
    by = c("drug_all" = "name_norm")
  )

cmdps_mapped_unique <- cmdps_mapped %>%
  group_by(drug_norm) %>%
  arrange(source_collapsed, .by_group = TRUE) %>%
  slice_head(n = 1) %>%
  ungroup() %>%
  distinct(lspci_id, drug_name = drug, drug_norm)

condition_vars <- c("cells", "stim", "stim_conc", "drug_id", "drug_conc", "time")

meta_cmpd_mapped <- meta_all %>% {
  bind_rows(
    filter(., is.na(lspci_id)) %>%
      select(-lspci_id) %>%
      left_join(
        cmdps_mapped_unique,
        by = "drug_norm"
      ),
    filter(., !is.na(lspci_id))
  )
} %>%
  mutate(
    # drug_id is either lspci_id or normalized drug name, if lspci_id was not found
    drug_id = if_else(!is.na(lspci_id), as.character(lspci_id), drug_norm)
  ) %>%
  mutate(
    condition = exec(
      paste,
      !!!as.list(.)[condition_vars],
      sep = "_"
    ),
    sample_id = paste(batch, well, condition, sep = "_")
  ) %>%
  group_nest(dataset, date, plate, batch, .key = "meta")

datasets <- datasets_counts %>%
  inner_join(meta_cmpd_mapped , by = c("dataset", "date", "plate")) %>%
  mutate_at(
    vars(counts, count_dupes, barcode_counts),
    map2,
    .$meta,
    ~.x %>%
      inner_join(
        distinct(.y, well, sample_id),
        by = "well"
      )
  ) %>%
  select(-meta)

library(qs)

qsave(
  datasets,
  file.path(wd, "counts_raw.qs"),
  preset = "high"
)

qsave(
  meta_cmpd_mapped,
  file.path(wd, "meta_mapped.qs"),
  preset = "high"
)

dataset_names <- tribble(
  ~dataset, ~dataset_name, ~date,
  "jsm_merck", "2017_09_john_santa_maria_merck_drugs", "2017_09",
  "lincs_cdk4_6_7", "2017_09_lincs_cdk4_6_inhibitors", "2017_09",
  "ld_dub", "2018_06_laura_doherty_dub_inhibitors", "2018_06",
  "lincs_cdk4_6_7", "2019_08_lincs_cdk4_6_7_inhibitors", "2019_08",
  "sr_repurposing", "2018_02_steven_rodriguez_ad_repurposing", "2018_02",
  "sr_repurposing", "2018_11_steven_rodriguez_ad_repurposing", "2018_11",
  "fp_transdiff", "2015_10_feodor_price_transdifferentiation_screen", "2015_10",
  "okl", "2021_08_optimal_kinase_library", "2021_08",
  "mixed_dge", "2023_02_ren-ovca-mel-test", "2023_02",
  "albers_dge", "2023_04_red-ad-repurposing-albers", "2023_04"
) %>%
  mutate(
    dataset_name = as.factor(dataset_name) %>%
      fct_inorder()
  )


write_csv(
  dataset_names,
  file.path(wd, "dataset_names.csv")
)


```

```{r synapse}

c(
  file.path(wd, "counts_raw.qs"),
  file.path(wd, "dataset_names.csv"),
  file.path(wd, "meta_mapped.qs")
) %>%
  synStoreMany(
    parentId = "syn21542764", forceVersion = FALSE,
    activityName = "Wrangle DGE data",
    used = c(
      "syn12979100",
      "syn21541012",
      "syn21542403",
      "syn21542711",
      "syn21586544",
      "syn24181333",
      "syn24183248",
      "syn25764541",
      dataset_syn %>%
        select(-dataset, -folder_syn, -plate, -date) %>%
        reduce(union)
    ),
    executed = "https://github.com/clemenshug/dge_comparison/blob/master/wrangling/09_dge_datasets.Rmd"
  )

#rsync -rv --exclude *.fq.gz --exclude fastq --exclude *.fastq.gz --exclude *.bam --exclude rapmap --exclude transcriptome 2018_02_steven_rodriguez_ad_repurposing transfer:/n/files/ImStor/sorger/data/rnaseq/lincs/dge_reanalysis_clemens/
```

